{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction of olfactory receptor (OR) responses to odorants\n",
    "\n",
    "This notebook trains `Random Forest` classification models to predict the response of human and mouse ORs to four odorants.   \n",
    "This is done in four steps:  \n",
    "1. load & transform training data;\n",
    "2. train 5 independent models with 5 subsets of residues, and compare their predictivity with a non-predictive control model;\n",
    "3. build a final model with the best residue subset above, then predict and rank new ORs by their probability to respond to the odorants;\n",
    "4. evaluate the prediction with *in vitro* data.\n",
    "\n",
    "The training data are **mostly** from the *in vitro* assays in the following publications:\n",
    "[Saito et al. Sci Signal (2009) 2:ra9](https://stke.sciencemag.org/content/2/60/ra9), \n",
    "[Mainland et al. Nat Neurosci (2014) 17:114–120](https://www.nature.com/articles/nn.3598#citeas), \n",
    "[Mainland et al. Sci Data (2015) 2:150002](https://www.nature.com/articles/sdata20152), \n",
    "[Jiang et al. Nat Neurosci (2015) 18:1446-1454](https://www.nature.com/articles/nn.4104), \n",
    "[Kida et al. Nat Commun (2018) 9:4556](https://www.nature.com/articles/s41467-018-06806-w), \n",
    "[Jones et al. Cell Syst (2019) 8:254-260](https://doi.org/10.1016/j.cels.2019.02.009).\n",
    "\n",
    "## Software requirements\n",
    "\n",
    "- `R` version 3.5 or higher is recommended.  \n",
    "- Required R packages: `caret`, `MLmetrics`, `e1071`, `bio3d`, `stringr`, `ggpubr`, `gridExtra`  \n",
    "It is recommended to install the packages with `Conda` (e.g. *conda install r-MLmetrics*, *conda install -c conda-forge r-ggpubr*) prior to using this notebook.  \n",
    "Alternative package installtion by `install.packages` (see below) may **fail**.  \n",
    "\n",
    "### Install & load required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: caret\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: ROCR\n",
      "\n",
      "Loading required package: MLmetrics\n",
      "\n",
      "\n",
      "Attaching package: ‘MLmetrics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:caret’:\n",
      "\n",
      "    MAE, RMSE\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    Recall\n",
      "\n",
      "\n",
      "Loading required package: e1071\n",
      "\n",
      "Loading required package: bio3d\n",
      "\n",
      "Loading required package: stringr\n",
      "\n",
      "Loading required package: ggpubr\n",
      "\n",
      "Loading required package: gridExtra\n",
      "\n"
     ]
    }
   ],
   "source": [
    "packages <- c(\"caret\", \"ROCR\", \"MLmetrics\", \"e1071\", \"bio3d\", \"stringr\", \"ggpubr\", \"gridExtra\")\n",
    "missing <- !unlist(lapply(packages, require, character.only = TRUE))\n",
    "\n",
    "if (any(missing)){\n",
    "    install.packages(packages[missing], dependencies = TRUE, repos='http://cran.us.r-project.org')\n",
    "    lapply((packages[missing]), library, character.only = TRUE)\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Training data  \n",
    "- **OR data**: pre-aligned sequences of human & mouse ORs, in which each amino acid will be converted to three physicochemical scores (`Composition`, `Polarity`, `Volumn`);  \n",
    "- **Odorant data**: 244 odorants represented by molecular descriptors, which have been pre-calculated with [Dragon 7.0](https://chm.kode-solutions.net/products_dragon.php);  \n",
    "- Finally, merge the above data into a **proteochemometric** presentation of OR-odorant pairs, labeled by **positive** (if the OR responds to the odorant *in vitro* in a dose-dependent manner) or **negative** (no significant response observed *in vitro*). \n",
    "\n",
    "### 1.1 Convert OR sequence alignment to physicochemical features\n",
    "\n",
    "Load the sequence alignment of human and mouse ORs, including some mutants. For this study, only the 7-transmembrane domain (7TM) and the extracellular loop 2 (ECL2) are included in the alignment, which are numbered according to the Ballesteros-Weinstein scheme. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ballesteros-Weinstein numbering scheme\n",
    "TM_ECL2 <- c(seq(1.32, 1.65, 0.01), # TM1\n",
    "             seq(2.39, 2.67, 0.01), # TM2\n",
    "             seq(3.25, 3.55, 0.01), # TM3\n",
    "             seq(4.40, 4.61, 0.01), # TM4\n",
    "             paste(\"EClII\", 1:23, sep=\"\"), # ECL2\n",
    "             seq(5.37, 5.65, 0.01), # TM5\n",
    "             seq(6.32, 6.56, 0.01), # TM6\n",
    "             seq(7.35, 7.55, 0.01)) # TM7\n",
    "\n",
    "seq <- read.fasta(\"Response_to_odorants/Data/Sequence_alignment.fa\")\n",
    "datOR <- as.data.frame(seq$ali)\n",
    "colnames(datOR) <- paste0(\"Res\", TM_ECL2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Take a look at the first 5 residues of 3 ORs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Res1.32</th><th scope=col>Res1.33</th><th scope=col>Res1.34</th><th scope=col>Res1.35</th><th scope=col>Res1.36</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>hOR10A2</th><td>Q</td><td>S</td><td>L</td><td>L</td><td>F</td></tr>\n",
       "\t<tr><th scope=row>hOR10A3</th><td>Q</td><td>V</td><td>Q</td><td>L</td><td>F</td></tr>\n",
       "\t<tr><th scope=row>hOR10A4</th><td>Q</td><td>A</td><td>L</td><td>L</td><td>F</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & Res1.32 & Res1.33 & Res1.34 & Res1.35 & Res1.36\\\\\n",
       "  & <fct> & <fct> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\thOR10A2 & Q & S & L & L & F\\\\\n",
       "\thOR10A3 & Q & V & Q & L & F\\\\\n",
       "\thOR10A4 & Q & A & L & L & F\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 5\n",
       "\n",
       "| <!--/--> | Res1.32 &lt;fct&gt; | Res1.33 &lt;fct&gt; | Res1.34 &lt;fct&gt; | Res1.35 &lt;fct&gt; | Res1.36 &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| hOR10A2 | Q | S | L | L | F |\n",
       "| hOR10A3 | Q | V | Q | L | F |\n",
       "| hOR10A4 | Q | A | L | L | F |\n",
       "\n"
      ],
      "text/plain": [
       "        Res1.32 Res1.33 Res1.34 Res1.35 Res1.36\n",
       "hOR10A2 Q       S       L       L       F      \n",
       "hOR10A3 Q       V       Q       L       F      \n",
       "hOR10A4 Q       A       L       L       F      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "datOR[1:3, 1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gaps from the alignment, then convert each amimo acid to its physicochemical scores (features): `Composition`, `Polarity`, `Volumn` [Grantham, Science (1974) 185:862-4](https://science.sciencemag.org/content/185/4154/862)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first 5 residues of 3 ORs after conversion:"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Res1.32Comp</th><th scope=col>Res1.32Polar</th><th scope=col>Res1.32Vol</th><th scope=col>Res1.34Comp</th><th scope=col>Res1.34Polar</th><th scope=col>Res1.34Vol</th><th scope=col>Res1.35Comp</th><th scope=col>Res1.35Polar</th><th scope=col>Res1.35Vol</th><th scope=col>Res1.36Comp</th><th scope=col>Res1.36Polar</th><th scope=col>Res1.36Vol</th><th scope=col>Res1.37Comp</th><th scope=col>Res1.37Polar</th><th scope=col>Res1.37Vol</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>hOR10A2</th><td>0.89</td><td>10.5</td><td>85</td><td>0.00</td><td> 4.9</td><td>111</td><td>0</td><td>4.9</td><td>111</td><td>0</td><td>5.2</td><td>132</td><td>0.00</td><td>4.9</td><td>111</td></tr>\n",
       "\t<tr><th scope=row>hOR10A3</th><td>0.89</td><td>10.5</td><td>85</td><td>0.89</td><td>10.5</td><td> 85</td><td>0</td><td>4.9</td><td>111</td><td>0</td><td>5.2</td><td>132</td><td>0.74</td><td>9.0</td><td>  3</td></tr>\n",
       "\t<tr><th scope=row>hOR10A4</th><td>0.89</td><td>10.5</td><td>85</td><td>0.00</td><td> 4.9</td><td>111</td><td>0</td><td>4.9</td><td>111</td><td>0</td><td>5.2</td><td>132</td><td>0.00</td><td>4.9</td><td>111</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & Res1.32Comp & Res1.32Polar & Res1.32Vol & Res1.34Comp & Res1.34Polar & Res1.34Vol & Res1.35Comp & Res1.35Polar & Res1.35Vol & Res1.36Comp & Res1.36Polar & Res1.36Vol & Res1.37Comp & Res1.37Polar & Res1.37Vol\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\thOR10A2 & 0.89 & 10.5 & 85 & 0.00 &  4.9 & 111 & 0 & 4.9 & 111 & 0 & 5.2 & 132 & 0.00 & 4.9 & 111\\\\\n",
       "\thOR10A3 & 0.89 & 10.5 & 85 & 0.89 & 10.5 &  85 & 0 & 4.9 & 111 & 0 & 5.2 & 132 & 0.74 & 9.0 &   3\\\\\n",
       "\thOR10A4 & 0.89 & 10.5 & 85 & 0.00 &  4.9 & 111 & 0 & 4.9 & 111 & 0 & 5.2 & 132 & 0.00 & 4.9 & 111\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 15\n",
       "\n",
       "| <!--/--> | Res1.32Comp &lt;dbl&gt; | Res1.32Polar &lt;dbl&gt; | Res1.32Vol &lt;dbl&gt; | Res1.34Comp &lt;dbl&gt; | Res1.34Polar &lt;dbl&gt; | Res1.34Vol &lt;dbl&gt; | Res1.35Comp &lt;dbl&gt; | Res1.35Polar &lt;dbl&gt; | Res1.35Vol &lt;dbl&gt; | Res1.36Comp &lt;dbl&gt; | Res1.36Polar &lt;dbl&gt; | Res1.36Vol &lt;dbl&gt; | Res1.37Comp &lt;dbl&gt; | Res1.37Polar &lt;dbl&gt; | Res1.37Vol &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| hOR10A2 | 0.89 | 10.5 | 85 | 0.00 |  4.9 | 111 | 0 | 4.9 | 111 | 0 | 5.2 | 132 | 0.00 | 4.9 | 111 |\n",
       "| hOR10A3 | 0.89 | 10.5 | 85 | 0.89 | 10.5 |  85 | 0 | 4.9 | 111 | 0 | 5.2 | 132 | 0.74 | 9.0 |   3 |\n",
       "| hOR10A4 | 0.89 | 10.5 | 85 | 0.00 |  4.9 | 111 | 0 | 4.9 | 111 | 0 | 5.2 | 132 | 0.00 | 4.9 | 111 |\n",
       "\n"
      ],
      "text/plain": [
       "        Res1.32Comp Res1.32Polar Res1.32Vol Res1.34Comp Res1.34Polar Res1.34Vol\n",
       "hOR10A2 0.89        10.5         85         0.00         4.9         111       \n",
       "hOR10A3 0.89        10.5         85         0.89        10.5          85       \n",
       "hOR10A4 0.89        10.5         85         0.00         4.9         111       \n",
       "        Res1.35Comp Res1.35Polar Res1.35Vol Res1.36Comp Res1.36Polar Res1.36Vol\n",
       "hOR10A2 0           4.9          111        0           5.2          132       \n",
       "hOR10A3 0           4.9          111        0           5.2          132       \n",
       "hOR10A4 0           4.9          111        0           5.2          132       \n",
       "        Res1.37Comp Res1.37Polar Res1.37Vol\n",
       "hOR10A2 0.00        4.9          111       \n",
       "hOR10A3 0.74        9.0            3       \n",
       "hOR10A4 0.00        4.9          111       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove columns that contain gaps in > 1% of the ORs \n",
    "datOR <- datOR[, colSums(datOR == \"-\") < nrow(datOR)*0.01]\n",
    "\n",
    "# Remove ORs that still contain gaps\n",
    "datOR <- datOR[rowSums(datOR == \"-\") == 0, ]\n",
    "\n",
    "# Convert amino acids to physicochemical features\n",
    "source(\"Functions/sequence2features_Grantham1974.R\")\n",
    "datOR <- cbind(residProps(datOR, \"Vol\"),\n",
    "               residProps(datOR, \"Comp\"),\n",
    "               residProps(datOR, \"Polar\"))\n",
    "datOR <- datOR[, sort(names(datOR))]\n",
    "\n",
    "cat(\"The first 5 residues of 3 ORs after conversion:\")\n",
    "datOR[1:3,1:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The alignment now contains 1727 ORs and 398 features."
     ]
    }
   ],
   "source": [
    "# Remove low variance features\n",
    "near_zero <- nearZeroVar(datOR[, ], freqCut = 95/5, saveMetrics = TRUE)\n",
    "low_variance_cols <- near_zero[(near_zero$zeroVar == TRUE) | (near_zero$nzv == TRUE), ]\n",
    "drops <- rownames(low_variance_cols)\n",
    "datOR <- datOR[, !(names(datOR) %in% drops)]\n",
    "\n",
    "# Remove highly correlated features\n",
    "correlationMatrix <- cor(datOR)\n",
    "highlyCorrelated <- findCorrelation(correlationMatrix, cutoff = 0.8)\n",
    "datOR <- datOR[, -highlyCorrelated]\n",
    "\n",
    "cat(\"The alignment now contains\", nrow(datOR), \"ORs and\", ncol(datOR), \"features.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The training set contains 708 ORs, including 318 orphan ORs."
     ]
    }
   ],
   "source": [
    "# Extract training set ORs\n",
    "trainOR <- read.csv('Response_to_odorants/Data/TrainingSet_OR-odorant_pairs/trainORs_labeled.csv',\n",
    "                    stringsAsFactors = FALSE)\n",
    "cat(\"The training set contains\", nrow(trainOR), \"ORs, including\", sum(trainOR$Label == \"Orphan\"), \"orphan ORs.\")\n",
    "\n",
    "datTrainOR <- datOR[rownames(datOR) %in% trainOR$OR, ]\n",
    "trainOR <- trainOR[trainOR$OR %in% rownames(datTrainOR), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Molecular descriptors of odorants  \n",
    "  \n",
    "Load pre-calculated molecular descriptors (features) of 244 odorants.  \n",
    "Each odorant has a unique PubChem CID, a SMILES string and 3850 features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Odorant</th><th scope=col>CID</th><th scope=col>SMILES</th><th scope=col>Feat1</th><th scope=col>Feat2</th><th scope=col>Feat3</th><th scope=col>Feat4</th><th scope=col>Feat5</th><th scope=col>Feat6</th><th scope=col>Feat7</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>phenylmethanol</td><td>244 </td><td>OCc1ccccc1         </td><td>108.15</td><td>6.759</td><td> 9.822</td><td>15.862</td><td>10.500</td><td>17.870</td><td>0.614</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>benzaldehyde  </td><td>240 </td><td>O=Cc1ccccc1        </td><td>106.13</td><td>7.581</td><td> 9.295</td><td>13.978</td><td> 9.739</td><td>15.455</td><td>0.664</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Diphenyl ether</td><td>7583</td><td>O(c1ccccc1)c2ccccc2</td><td>170.22</td><td>7.401</td><td>15.349</td><td>22.745</td><td>16.262</td><td>25.285</td><td>0.667</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & Odorant & CID & SMILES & Feat1 & Feat2 & Feat3 & Feat4 & Feat5 & Feat6 & Feat7\\\\\n",
       "  & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & phenylmethanol & 244  & OCc1ccccc1          & 108.15 & 6.759 &  9.822 & 15.862 & 10.500 & 17.870 & 0.614\\\\\n",
       "\t2 & benzaldehyde   & 240  & O=Cc1ccccc1         & 106.13 & 7.581 &  9.295 & 13.978 &  9.739 & 15.455 & 0.664\\\\\n",
       "\t3 & Diphenyl ether & 7583 & O(c1ccccc1)c2ccccc2 & 170.22 & 7.401 & 15.349 & 22.745 & 16.262 & 25.285 & 0.667\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 10\n",
       "\n",
       "| <!--/--> | Odorant &lt;chr&gt; | CID &lt;chr&gt; | SMILES &lt;chr&gt; | Feat1 &lt;dbl&gt; | Feat2 &lt;dbl&gt; | Feat3 &lt;dbl&gt; | Feat4 &lt;dbl&gt; | Feat5 &lt;dbl&gt; | Feat6 &lt;dbl&gt; | Feat7 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | phenylmethanol | 244  | OCc1ccccc1          | 108.15 | 6.759 |  9.822 | 15.862 | 10.500 | 17.870 | 0.614 |\n",
       "| 2 | benzaldehyde   | 240  | O=Cc1ccccc1         | 106.13 | 7.581 |  9.295 | 13.978 |  9.739 | 15.455 | 0.664 |\n",
       "| 3 | Diphenyl ether | 7583 | O(c1ccccc1)c2ccccc2 | 170.22 | 7.401 | 15.349 | 22.745 | 16.262 | 25.285 | 0.667 |\n",
       "\n"
      ],
      "text/plain": [
       "  Odorant        CID  SMILES              Feat1  Feat2 Feat3  Feat4  Feat5 \n",
       "1 phenylmethanol 244  OCc1ccccc1          108.15 6.759  9.822 15.862 10.500\n",
       "2 benzaldehyde   240  O=Cc1ccccc1         106.13 7.581  9.295 13.978  9.739\n",
       "3 Diphenyl ether 7583 O(c1ccccc1)c2ccccc2 170.22 7.401 15.349 22.745 16.262\n",
       "  Feat6  Feat7\n",
       "1 17.870 0.614\n",
       "2 15.455 0.664\n",
       "3 25.285 0.667"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allOdor <- read.csv(\"Response_to_odorants/Data/Odorants_features.csv\", header = 1,\n",
    "                    colClasses=c('character', 'character', 'character', replicate(3850, 'numeric')))\n",
    "allOdor[1:3, 1:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The training set will contain the 4 odorants of interest, as well as their analogs.  \n",
    "To to find their analogs, we use a pre-calculated chemical similarity (`Tanimono coefficient`) matrix between the 244 odorants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The training set contains: \n",
      "\n",
      " 4 odorants of interest"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 4 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Odorant</th><th scope=col>CID</th><th scope=col>SMILES</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>128</th><td>4-chromanone </td><td>68110 </td><td>O=C1CCOc2ccccc12        </td></tr>\n",
       "\t<tr><th scope=row>186</th><td>coumarin     </td><td>323   </td><td>O=C1Oc2ccccc2C=C1       </td></tr>\n",
       "\t<tr><th scope=row>197</th><td>acetophenone </td><td>7410  </td><td>CC(=O)c1ccccc1          </td></tr>\n",
       "\t<tr><th scope=row>223</th><td>R-(-)-Carvone</td><td>439570</td><td>CC1=CC[C@H](CC1=O)C(=C)C</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & Odorant & CID & SMILES\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t128 & 4-chromanone  & 68110  & O=C1CCOc2ccccc12        \\\\\n",
       "\t186 & coumarin      & 323    & O=C1Oc2ccccc2C=C1       \\\\\n",
       "\t197 & acetophenone  & 7410   & CC(=O)c1ccccc1          \\\\\n",
       "\t223 & R-(-)-Carvone & 439570 & CC1=CC{[}C@H{]}(CC1=O)C(=C)C\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 3\n",
       "\n",
       "| <!--/--> | Odorant &lt;chr&gt; | CID &lt;chr&gt; | SMILES &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 128 | 4-chromanone  | 68110  | O=C1CCOc2ccccc12         |\n",
       "| 186 | coumarin      | 323    | O=C1Oc2ccccc2C=C1        |\n",
       "| 197 | acetophenone  | 7410   | CC(=O)c1ccccc1           |\n",
       "| 223 | R-(-)-Carvone | 439570 | CC1=CC[C@H](CC1=O)C(=C)C |\n",
       "\n"
      ],
      "text/plain": [
       "    Odorant       CID    SMILES                  \n",
       "128 4-chromanone  68110  O=C1CCOc2ccccc12        \n",
       "186 coumarin      323    O=C1Oc2ccccc2C=C1       \n",
       "197 acetophenone  7410   CC(=O)c1ccccc1          \n",
       "223 R-(-)-Carvone 439570 CC1=CC[C@H](CC1=O)C(=C)C"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "and their 6 analogs:"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Odorant</th><th scope=col>CID</th><th scope=col>SMILES</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>98</th><td>S-(+)-carvone    </td><td>16724   </td><td>CC1=CC[C@@H](CC1=O)C(=C)C  </td></tr>\n",
       "\t<tr><th scope=row>144</th><td>(R)-(+)-limonene </td><td>440917  </td><td>CC1=CC[C@@H](CC1)C(=C)C    </td></tr>\n",
       "\t<tr><th scope=row>145</th><td>(S)-(-)-limonene </td><td>439250  </td><td>CC1=CC[C@H](CC1)C(=C)C     </td></tr>\n",
       "\t<tr><th scope=row>157</th><td>Isolimonene      </td><td>22831540</td><td>CC(=C)C1CCC(=CC1)C         </td></tr>\n",
       "\t<tr><th scope=row>201</th><td>methyl benzoate  </td><td>7150    </td><td>COC(=O)C1=CC=CC=C1         </td></tr>\n",
       "\t<tr><th scope=row>243</th><td>4-Hydroxycoumarin</td><td>54682930</td><td>C1=CC=C2C(=C1)C(=CC(=O)O2)O</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & Odorant & CID & SMILES\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t98 & S-(+)-carvone     & 16724    & CC1=CC{[}C@@H{]}(CC1=O)C(=C)C  \\\\\n",
       "\t144 & (R)-(+)-limonene  & 440917   & CC1=CC{[}C@@H{]}(CC1)C(=C)C    \\\\\n",
       "\t145 & (S)-(-)-limonene  & 439250   & CC1=CC{[}C@H{]}(CC1)C(=C)C     \\\\\n",
       "\t157 & Isolimonene       & 22831540 & CC(=C)C1CCC(=CC1)C         \\\\\n",
       "\t201 & methyl benzoate   & 7150     & COC(=O)C1=CC=CC=C1         \\\\\n",
       "\t243 & 4-Hydroxycoumarin & 54682930 & C1=CC=C2C(=C1)C(=CC(=O)O2)O\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | Odorant &lt;chr&gt; | CID &lt;chr&gt; | SMILES &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 98 | S-(+)-carvone     | 16724    | CC1=CC[C@@H](CC1=O)C(=C)C   |\n",
       "| 144 | (R)-(+)-limonene  | 440917   | CC1=CC[C@@H](CC1)C(=C)C     |\n",
       "| 145 | (S)-(-)-limonene  | 439250   | CC1=CC[C@H](CC1)C(=C)C      |\n",
       "| 157 | Isolimonene       | 22831540 | CC(=C)C1CCC(=CC1)C          |\n",
       "| 201 | methyl benzoate   | 7150     | COC(=O)C1=CC=CC=C1          |\n",
       "| 243 | 4-Hydroxycoumarin | 54682930 | C1=CC=C2C(=C1)C(=CC(=O)O2)O |\n",
       "\n"
      ],
      "text/plain": [
       "    Odorant           CID      SMILES                     \n",
       "98  S-(+)-carvone     16724    CC1=CC[C@@H](CC1=O)C(=C)C  \n",
       "144 (R)-(+)-limonene  440917   CC1=CC[C@@H](CC1)C(=C)C    \n",
       "145 (S)-(-)-limonene  439250   CC1=CC[C@H](CC1)C(=C)C     \n",
       "157 Isolimonene       22831540 CC(=C)C1CCC(=CC1)C         \n",
       "201 methyl benzoate   7150     COC(=O)C1=CC=CC=C1         \n",
       "243 4-Hydroxycoumarin 54682930 C1=CC=C2C(=C1)C(=CC(=O)O2)O"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load the chemical similarity matrix\n",
    "tanimoto <- read.csv('Response_to_odorants/Data/Odorants_similarity.csv', row.names = 1)\n",
    "\n",
    "# PubChem CIDs of the 4 odorants of interest\n",
    "testCIDs <- c(439570, 7410, 323, 68110)\n",
    "\n",
    "# Tanimoto coeeficient > 0.6 are considered similar\n",
    "numextract <- function(string){ \n",
    "  as.numeric(str_extract(string, \"\\\\-*\\\\d+\\\\.*\\\\d*\"))\n",
    "}\n",
    "\n",
    "trainCIDs <- numextract(colnames(tanimoto)[colSums(tanimoto[paste0('CID', testCIDs), ] > 0.6) > 0])\n",
    "\n",
    "cat(\"The training set contains: \\n\\n\", length(testCIDs), \"odorants of interest\")\n",
    "allOdor[allOdor$CID %in% testCIDs, 1:3]\n",
    "\n",
    "cat(\"and their\", length(setdiff(trainCIDs, testCIDs)), \"analogs:\")\n",
    "allOdor[allOdor$CID %in% setdiff(trainCIDs, testCIDs), 1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 features remained:"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 10 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>CID</th><th scope=col>Feat805</th><th scope=col>Feat824</th><th scope=col>Feat1032</th><th scope=col>Feat1066</th><th scope=col>Feat1099</th><th scope=col>Feat1152</th><th scope=col>Feat1517</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>98</th><td>16724   </td><td>1.278</td><td>0.004</td><td> 0.044</td><td>1.108</td><td>0.049</td><td>3.552</td><td>0.68</td></tr>\n",
       "\t<tr><th scope=row>128</th><td>68110   </td><td>1.258</td><td>0.017</td><td>-0.057</td><td>1.141</td><td>0.046</td><td>3.370</td><td>0.86</td></tr>\n",
       "\t<tr><th scope=row>144</th><td>440917  </td><td>1.272</td><td>0.026</td><td> 0.091</td><td>1.071</td><td>0.045</td><td>3.476</td><td>0.68</td></tr>\n",
       "\t<tr><th scope=row>145</th><td>439250  </td><td>1.272</td><td>0.026</td><td> 0.091</td><td>1.071</td><td>0.045</td><td>3.476</td><td>0.68</td></tr>\n",
       "\t<tr><th scope=row>157</th><td>22831540</td><td>1.272</td><td>0.026</td><td> 0.091</td><td>1.071</td><td>0.045</td><td>3.476</td><td>0.68</td></tr>\n",
       "\t<tr><th scope=row>186</th><td>323     </td><td>1.279</td><td>0.018</td><td> 0.153</td><td>1.063</td><td>0.031</td><td>3.463</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>197</th><td>7410    </td><td>1.286</td><td>0.015</td><td> 0.030</td><td>1.233</td><td>0.049</td><td>3.461</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>201</th><td>7150    </td><td>1.236</td><td>0.011</td><td> 0.080</td><td>0.941</td><td>0.040</td><td>3.555</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>223</th><td>439570  </td><td>1.278</td><td>0.004</td><td> 0.044</td><td>1.108</td><td>0.049</td><td>3.552</td><td>0.68</td></tr>\n",
       "\t<tr><th scope=row>243</th><td>54682930</td><td>1.250</td><td>0.026</td><td> 0.093</td><td>1.111</td><td>0.059</td><td>3.779</td><td>0.00</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & CID & Feat805 & Feat824 & Feat1032 & Feat1066 & Feat1099 & Feat1152 & Feat1517\\\\\n",
       "  & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t98 & 16724    & 1.278 & 0.004 &  0.044 & 1.108 & 0.049 & 3.552 & 0.68\\\\\n",
       "\t128 & 68110    & 1.258 & 0.017 & -0.057 & 1.141 & 0.046 & 3.370 & 0.86\\\\\n",
       "\t144 & 440917   & 1.272 & 0.026 &  0.091 & 1.071 & 0.045 & 3.476 & 0.68\\\\\n",
       "\t145 & 439250   & 1.272 & 0.026 &  0.091 & 1.071 & 0.045 & 3.476 & 0.68\\\\\n",
       "\t157 & 22831540 & 1.272 & 0.026 &  0.091 & 1.071 & 0.045 & 3.476 & 0.68\\\\\n",
       "\t186 & 323      & 1.279 & 0.018 &  0.153 & 1.063 & 0.031 & 3.463 & 0.00\\\\\n",
       "\t197 & 7410     & 1.286 & 0.015 &  0.030 & 1.233 & 0.049 & 3.461 & 0.00\\\\\n",
       "\t201 & 7150     & 1.236 & 0.011 &  0.080 & 0.941 & 0.040 & 3.555 & 0.00\\\\\n",
       "\t223 & 439570   & 1.278 & 0.004 &  0.044 & 1.108 & 0.049 & 3.552 & 0.68\\\\\n",
       "\t243 & 54682930 & 1.250 & 0.026 &  0.093 & 1.111 & 0.059 & 3.779 & 0.00\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 8\n",
       "\n",
       "| <!--/--> | CID &lt;fct&gt; | Feat805 &lt;dbl&gt; | Feat824 &lt;dbl&gt; | Feat1032 &lt;dbl&gt; | Feat1066 &lt;dbl&gt; | Feat1099 &lt;dbl&gt; | Feat1152 &lt;dbl&gt; | Feat1517 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 98 | 16724    | 1.278 | 0.004 |  0.044 | 1.108 | 0.049 | 3.552 | 0.68 |\n",
       "| 128 | 68110    | 1.258 | 0.017 | -0.057 | 1.141 | 0.046 | 3.370 | 0.86 |\n",
       "| 144 | 440917   | 1.272 | 0.026 |  0.091 | 1.071 | 0.045 | 3.476 | 0.68 |\n",
       "| 145 | 439250   | 1.272 | 0.026 |  0.091 | 1.071 | 0.045 | 3.476 | 0.68 |\n",
       "| 157 | 22831540 | 1.272 | 0.026 |  0.091 | 1.071 | 0.045 | 3.476 | 0.68 |\n",
       "| 186 | 323      | 1.279 | 0.018 |  0.153 | 1.063 | 0.031 | 3.463 | 0.00 |\n",
       "| 197 | 7410     | 1.286 | 0.015 |  0.030 | 1.233 | 0.049 | 3.461 | 0.00 |\n",
       "| 201 | 7150     | 1.236 | 0.011 |  0.080 | 0.941 | 0.040 | 3.555 | 0.00 |\n",
       "| 223 | 439570   | 1.278 | 0.004 |  0.044 | 1.108 | 0.049 | 3.552 | 0.68 |\n",
       "| 243 | 54682930 | 1.250 | 0.026 |  0.093 | 1.111 | 0.059 | 3.779 | 0.00 |\n",
       "\n"
      ],
      "text/plain": [
       "    CID      Feat805 Feat824 Feat1032 Feat1066 Feat1099 Feat1152 Feat1517\n",
       "98  16724    1.278   0.004    0.044   1.108    0.049    3.552    0.68    \n",
       "128 68110    1.258   0.017   -0.057   1.141    0.046    3.370    0.86    \n",
       "144 440917   1.272   0.026    0.091   1.071    0.045    3.476    0.68    \n",
       "145 439250   1.272   0.026    0.091   1.071    0.045    3.476    0.68    \n",
       "157 22831540 1.272   0.026    0.091   1.071    0.045    3.476    0.68    \n",
       "186 323      1.279   0.018    0.153   1.063    0.031    3.463    0.00    \n",
       "197 7410     1.286   0.015    0.030   1.233    0.049    3.461    0.00    \n",
       "201 7150     1.236   0.011    0.080   0.941    0.040    3.555    0.00    \n",
       "223 439570   1.278   0.004    0.044   1.108    0.049    3.552    0.68    \n",
       "243 54682930 1.250   0.026    0.093   1.111    0.059    3.779    0.00    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract the molecular features of the training set odorants.  \n",
    "datOdor <- allOdor[allOdor$CID %in% trainCIDs, ]\n",
    "CID <- datOdor$CID\n",
    "datOdor <- datOdor[, -c(1:3)]\n",
    "\n",
    "# Removes columns containing NA\n",
    "datOdor <- datOdor[, colSums(is.na(datOdor)) == 0]\n",
    "\n",
    "# Remove low variance features\n",
    "near_zero <- nearZeroVar(datOdor[, ], freqCut = 95/5, saveMetrics = TRUE)\n",
    "low_variance_cols <- near_zero[(near_zero$zeroVar == TRUE) | (near_zero$nzv == TRUE), ]\n",
    "drops <- rownames(low_variance_cols)\n",
    "datOdor <- datOdor[, !(names(datOdor) %in% drops)]\n",
    "\n",
    "# Remove highly correlated features\n",
    "correlationMatrix <- cor(datOdor)\n",
    "highlyCorrelated <- findCorrelation(correlationMatrix, cutoff=0.75)\n",
    "datOdor <- datOdor[, -highlyCorrelated]\n",
    "datOdor <- cbind(CID, datOdor)\n",
    "\n",
    "cat(ncol(datOdor), \"features remained:\")\n",
    "datOdor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Merge OR and odorant data into OR-odorant pairs  \n",
    "For each odorant, merge it with all the ORs that have been tested on it *in vitro*, then label each pair with **positive** or **negative**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The final training set contains 1065 OR-odorant pairs:"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "positive negative \n",
       "     202      863 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allPairs <- read.csv(\"Response_to_odorants/Data/TrainingSet_OR-odorant_pairs/OR-odorant_pairs.csv\",\n",
    "                     stringsAsFactors = FALSE)\n",
    "\n",
    "datPos <- data.frame()\n",
    "datNeg <- data.frame()\n",
    "\n",
    "for (cid in trainCIDs){\n",
    "    # merge responsive pairs\n",
    "    pos <- allPairs$Gene[allPairs$CID == cid & allPairs$Responsive == 1]\n",
    "    if (length(pos) > 0){\n",
    "        datP <- merge(datTrainOR[rownames(datTrainOR) %in% pos, ], datOdor[datOdor$CID == cid, -1])\n",
    "        rownames(datP) <- paste(rownames(datTrainOR)[rownames(datTrainOR) %in% pos], cid, sep = \"_\")\n",
    "        datPos <- rbind(datPos, datP)\n",
    "    }\n",
    "    \n",
    "    # merge non-responsive pairs\n",
    "    neg <- allPairs$Gene[allPairs$CID == cid & allPairs$Responsive == 0]\n",
    "    if (length(neg) > 0){\n",
    "        datN <- merge(datTrainOR[rownames(datTrainOR) %in% neg, ], datOdor[datOdor$CID == cid, -1])\n",
    "        rownames(datN) <- paste(rownames(datTrainOR)[rownames(datTrainOR) %in% neg], cid, sep = \"_\")\n",
    "        datNeg <- rbind(datNeg, datN)\n",
    "    }\n",
    "}\n",
    "\n",
    "# Add a column of class labels\n",
    "datPos$Label <- \"positive\"\n",
    "datNeg$Label <- \"negative\"\n",
    "\n",
    "data <- rbind(datPos, datNeg)\n",
    "data$Label <- factor(data$Label, levels = c(\"positive\", \"negative\"))\n",
    "\n",
    "cat(\"The final training set contains\", nrow(data), \"OR-odorant pairs:\")\n",
    "table(data$Label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Model training  \n",
    "Train 5 independent models, starting from 5 heuristic subsets of residues in the OR sequences:\n",
    "- ***poc17***, 17 residues outlining the odorant-binding pocket\n",
    "- ***poc20***, 20 residues outlining the odorant-binding pocket (*poc17* + 3 peripheral residues)\n",
    "- ***poc27***, *poc20* + 7 more peripheral residues\n",
    "- ***poc60***, 60 residues in an extended pocket region\n",
    "- ***TM191***, 191 residues in the transmembrane (TM) domain\n",
    "\n",
    "As control, a non-predictive model will be built with pseudo-sequences of 60 random residues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsets <- c('poc17', 'poc20', 'poc27', 'poc60', 'TM191', 'random')\n",
    "\n",
    "# Ballesteros-Weinstein numbers of 5 heuristic subsets of residues\n",
    "poc17 <- c(2.53,  # TM2\n",
    "           3.29, 3.30, 3.32, 3.33, 3.36, 3.37, 3.38, # TM3\n",
    "           5.42, 5.43, 5.46, 5.47, # TM5\n",
    "           6.48, 6.51, 6.52, 6.55, #TM6\n",
    "           7.42) # TM7\n",
    "\n",
    "poc20 <- c(2.53, 2.54, # TM2\n",
    "           3.29, 3.30, 3.32, 3.33, 3.36, 3.37, 3.38, # TM3\n",
    "           5.38, 5.39, 5.42, 5.43, 5.46, 5.47, # TM5\n",
    "           6.48, 6.51, 6.52, 6.55, #TM6\n",
    "           7.42) # TM7\n",
    "\n",
    "poc27 <- c(2.53, 2.54, # TM2\n",
    "           3.29, 3.30, 3.31, 3.32, 3.33, 3.36, 3.37, 3.38, 3.40, 3.41, # TM3\n",
    "           4.56, # TM4\n",
    "           5.37, 5.38, 5.39, 5.42, 5.43, 5.46, 5.47, # TM5\n",
    "           6.44, 6.48, 6.51, 6.52, 6.55, #TM6\n",
    "           7.38, 7.42) # TM7\n",
    "\n",
    "poc60 <- c(2.53,2.54, 2.57, 2.61, # TM2\n",
    "           seq(3.25, 3.41, 0.01), # TM3\n",
    "           4.52, 4.53, 4.56, 4.57, 4.60, # TM4\n",
    "           seq(5.35, 5.51, 0.01), # TM5\n",
    "           6.44, 6.45, seq(6.47, 6.56, 0.01), # TM6\n",
    "           7.35, 7.38, 7.39, 7.41, 7.42) # TM7\n",
    "\n",
    "TM191 <- c(seq(1.32, 1.65, 0.01), # TM1\n",
    "           seq(2.39, 2.67, 0.01), # TM2\n",
    "           seq(3.25, 3.55, 0.01), # TM3\n",
    "           seq(4.40, 4.61, 0.01), # TM4\n",
    "           seq(5.37, 5.65, 0.01), # TM5\n",
    "           seq(6.32, 6.56, 0.01), # TM6\n",
    "           seq(7.35, 7.55, 0.01)) # TM7\n",
    "\n",
    "# for the non-predictive control model\n",
    "random <- sample(TM191, 60)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 5-fold cross validation  \n",
    "**Note:** The validation set contains only the 4 odorants of interest and not their analogs. The final model will serve the 4 odorants and should not be evaluated on the analogs. The analogs are included to enhance the model training and and should remain in the training set. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Index the pairs containing the 4 odorants of interest\n",
    "val_idx <- unlist(lapply(paste0(\"_\", testCIDs), function(x) grep(x, rownames(data))))\n",
    "valPairs <- rownames(data)[val_idx]\n",
    "                         \n",
    "# Stratify into 5 folds (as validation sets) for cross validation\n",
    "folds <- createFolds(data[val_idx, ]$Label, 5, list = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract metrics from a confusion matrix\n",
    "getMetrics <- function(prediction, reference, positive){\n",
    "    # confusion matix\n",
    "    p <- confusionMatrix(prediction, reference, positive = positive)\n",
    "    \n",
    "    TP <- p$table[1]\n",
    "    TN <- p$table[4]\n",
    "    FP <- p$table[3]\n",
    "    FN <- p$table[2]\n",
    "\n",
    "    # Matthews correlation coefficient (MCC)\n",
    "    mcc_num <- (TP*TN - FP*FN)\n",
    "    mcc_den <- as.double((TP+FP))*as.double((TP+FN))*as.double((TN+FP))*as.double((TN+FN))\n",
    "    MCC <- mcc_num/sqrt(mcc_den)\n",
    "    \n",
    "    metrics=c(round(c(MCC, p$byClass[c(5,6,2)]), 2), TP, TN, FP, FN)\n",
    "    names(metrics)=c(\"MCC\", \"Precision\", \"Recall\", \"Specificity\", \"TP\", \"TN\", \"FP\", \"FN\")\n",
    "    return(metrics)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train 6 models with the 6 residue subsets\n",
    "for (subset in subsets){\n",
    "    \n",
    "    # Extract the data columns belonging to the residue subset, as well as the odorant columns and the response label \n",
    "    ResCols <- paste0(\"Res\", as.vector(outer(get(subset), c(\"Comp\", \"Polar\", \"Vol\"), paste0)))\n",
    "    datTrain <- data[, names(data) %in% c(ResCols, names(datOdor), \"Label\")]\n",
    "    \n",
    "    # for the control model, generate pseudo-sequences by randomizing the residues and the responses to the odorants\n",
    "    if (subset == \"random\"){\n",
    "        pseudoSeq <- data[, names(data) %in% ResCols]\n",
    "        pseudoSeq <- pseudoSeq[sample(nrow(pseudoSeq)), sample(ncol(pseudoSeq))]\n",
    "        datTrain <- cbind(pseudoSeq, data[, names(data) %in% c(names(datOdor), \"Label\")])\n",
    "    } \n",
    "    \n",
    "    # An empty matrix to summarize the metrics of model performance\n",
    "    metrics <- c(\"MCC\", \"Precision\", \"Recall\", \"Specificity\", \"TP\", \"TN\", \"FP\", \"FN\")\n",
    "    perfSum <- matrix(nrow = 5, ncol = length(metrics))\n",
    "    colnames(perfSum) <- metrics\n",
    "    rownames(perfSum) <- paste(\"Fold\", 1:5, sep=\"_\")\n",
    "    \n",
    "    # 5-fold cross validation\n",
    "    for (i in 1:5){\n",
    "        # Split training & validation sets\n",
    "        valSet <- datTrain[rownames(datTrain) %in% valPairs[unlist(folds[i])], ]\n",
    "        trainingSet <- datTrain[! rownames(datTrain) %in% valPairs[unlist(folds[i])], ]\n",
    "        \n",
    "        # Center & scale data\n",
    "        preProc <- preProcess(trainingSet, method = c(\"center\", \"scale\"))\n",
    "        trainingProc <- predict(preProc, trainingSet)\n",
    "        valProc <- predict(preProc, valSet)\n",
    "        \n",
    "        # Parameter tuning using nested 5-fold cross validation\n",
    "        control <- trainControl(classProbs = TRUE, \n",
    "                            method = \"cv\", \n",
    "                            number = 5,\n",
    "                            summaryFunction = prSummary)\n",
    "        grid <- expand.grid(.mtry = seq(ncol(trainingProc)/2, ncol(trainingProc)-1, length.out= 5)) \n",
    "        \n",
    "        # Train a random forest model with the tuned parameters\n",
    "        mod_RF <- train(data = trainingProc, Label ~ . , \n",
    "                        method=\"rf\", ntree = 2001,\n",
    "                        metric=\"Precision\",\n",
    "                        tuneGrid = grid,\n",
    "                        trControl=control)\n",
    "        \n",
    "        # Use the model to predicte the response probability of the OR-odorant pairs in the validation set\n",
    "        probability <- predict(mod_RF, valProc, type = \"prob\")\n",
    "        \n",
    "        # Classifiy probability > 0.5 as \"positive\", equivalent to predict(mod_fit, newdata = valProc)\n",
    "        predClass <- factor(ifelse(probability$positive > 0.5, \"positive\", \"negative\"),\n",
    "                              levels = c(\"positive\", \"negative\"))\n",
    "        \n",
    "        # Summarize the model performance on the validation set\n",
    "        perfSum[i,] <- getMetrics(predClass, valProc$Label, positive = \"positive\")\n",
    "    }\n",
    "    \n",
    "    # Mean, SD & SEM of the performance summary\n",
    "    perfSum <- rbind(perfSum, Mean = round(colMeans(perfSum), 2), SD =round(apply(perfSum, 2, sd), 2))\n",
    "    perfSum <- rbind(perfSum, SEM =round(perfSum[\"SD\", ]/sqrt(5), 2))\n",
    "    cat(\"\\nPerformance summary of model\", subset, \"\\n\")\n",
    "    print(perfSum)\n",
    "    \n",
    "    # Output  to file\n",
    "    write.table(perfSum, paste0(\"Response_to_odorants/Cross_validation/\", subset, \".tsv\"), sep = \"\\t\", quote = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Baseline: machine learning *vs* statistics\n",
    "How good are the machine learning models, really? How much can we get simply from statistical inferences?\n",
    "Assuming (naively) that the probability for an OR-odorant pair to be responsive is the average of the OR's responsiveness to the other odorants and the odorant's responsiveness to the other ORs. That is to say, that promiscuous ORs and odorants have high probabilities to response to one another. This probability draws a baseline to evaluate the performance of the above `Random Forest` models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First rows of the baseline:"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>OR</th><th scope=col>odorant</th><th scope=col>resp_pair</th><th scope=col>stat_prob</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>hOR10A6_179</th><td>hOR10A6                 </td><td>179</td><td>2</td><td>0.01</td></tr>\n",
       "\t<tr><th scope=row>hOR10A6_L287P_179</th><td>hOR10A6_L287P           </td><td>179</td><td>1</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>hOR10A6_V140P_L287P_179</th><td>hOR10A6_V140P_L287P     </td><td>179</td><td>1</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>hOR10C1_F60L_P174Q_M246V_179</th><td>hOR10C1_F60L_P174Q_M246V</td><td>179</td><td>1</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>hOR10G3_179</th><td>hOR10G3                 </td><td>179</td><td>2</td><td>0.01</td></tr>\n",
       "\t<tr><th scope=row>hOR10G3_S73G_179</th><td>hOR10G3_S73G            </td><td>179</td><td>4</td><td>0.04</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & OR & odorant & resp\\_pair & stat\\_prob\\\\\n",
       "  & <fct> & <fct> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\thOR10A6\\_179 & hOR10A6                  & 179 & 2 & 0.01\\\\\n",
       "\thOR10A6\\_L287P\\_179 & hOR10A6\\_L287P            & 179 & 1 & 0.00\\\\\n",
       "\thOR10A6\\_V140P\\_L287P\\_179 & hOR10A6\\_V140P\\_L287P      & 179 & 1 & 0.00\\\\\n",
       "\thOR10C1\\_F60L\\_P174Q\\_M246V\\_179 & hOR10C1\\_F60L\\_P174Q\\_M246V & 179 & 1 & 0.00\\\\\n",
       "\thOR10G3\\_179 & hOR10G3                  & 179 & 2 & 0.01\\\\\n",
       "\thOR10G3\\_S73G\\_179 & hOR10G3\\_S73G             & 179 & 4 & 0.04\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 4\n",
       "\n",
       "| <!--/--> | OR &lt;fct&gt; | odorant &lt;fct&gt; | resp_pair &lt;int&gt; | stat_prob &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| hOR10A6_179 | hOR10A6                  | 179 | 2 | 0.01 |\n",
       "| hOR10A6_L287P_179 | hOR10A6_L287P            | 179 | 1 | 0.00 |\n",
       "| hOR10A6_V140P_L287P_179 | hOR10A6_V140P_L287P      | 179 | 1 | 0.00 |\n",
       "| hOR10C1_F60L_P174Q_M246V_179 | hOR10C1_F60L_P174Q_M246V | 179 | 1 | 0.00 |\n",
       "| hOR10G3_179 | hOR10G3                  | 179 | 2 | 0.01 |\n",
       "| hOR10G3_S73G_179 | hOR10G3_S73G             | 179 | 4 | 0.04 |\n",
       "\n"
      ],
      "text/plain": [
       "                             OR                       odorant resp_pair\n",
       "hOR10A6_179                  hOR10A6                  179     2        \n",
       "hOR10A6_L287P_179            hOR10A6_L287P            179     1        \n",
       "hOR10A6_V140P_L287P_179      hOR10A6_V140P_L287P      179     1        \n",
       "hOR10C1_F60L_P174Q_M246V_179 hOR10C1_F60L_P174Q_M246V 179     1        \n",
       "hOR10G3_179                  hOR10G3                  179     2        \n",
       "hOR10G3_S73G_179             hOR10G3_S73G             179     4        \n",
       "                             stat_prob\n",
       "hOR10A6_179                  0.01     \n",
       "hOR10A6_L287P_179            0.00     \n",
       "hOR10A6_V140P_L287P_179      0.00     \n",
       "hOR10C1_F60L_P174Q_M246V_179 0.00     \n",
       "hOR10G3_179                  0.01     \n",
       "hOR10G3_S73G_179             0.04     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "..."
     ]
    }
   ],
   "source": [
    "# Responsiveness of each OR in the training set\n",
    "resp_OR <- table(allPairs$Gene[allPairs$Responsive==1])\n",
    "# Responsiveness of each odorant in the training set\n",
    "resp_odor <- table(allPairs$CID[allPairs$Responsive==1])\n",
    "# Average the responsiveness of each OR-odorant pair\n",
    "combi <- merge(names(resp_OR), names(resp_odor))\n",
    "names(combi) <- c(\"OR\", \"odorant\")\n",
    "for (i in 1:nrow(combi)){\n",
    "    combi$resp_pair[i] <- mean(resp_OR[names(resp_OR) == combi$OR[i]], resp_odor[names(resp_odor) == combi$odorant[i]])\n",
    "}\n",
    "\n",
    "# Normalize to obtain statistical probabilities\n",
    "norm01 <- function(x){(x-min(x))/(max(x)-min(x))} # a normalization function \n",
    "combi$stat_prob <- round(norm01(combi$resp_pair), 2)\n",
    "rownames(combi) <- paste(combi$OR, combi$odorant, sep = '_')\n",
    "cat(\"First rows of the baseline:\")\n",
    "head(combi)\n",
    "cat(\"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Performance summary of the baseline model\n",
      "        MCC Precision Recall Specificity   TP TN FP    FN\n",
      "Fold_1 0.16      1.00   0.03           1 1.00 76  0 28.00\n",
      "Fold_2 0.23      1.00   0.07           1 2.00 76  0 27.00\n",
      "Fold_3 0.28      1.00   0.10           1 3.00 76  0 26.00\n",
      "Fold_4 0.00      0.00   0.00           1 0.00 76  0 29.00\n",
      "Fold_5 0.16      1.00   0.03           1 1.00 76  0 28.00\n",
      "Mean   0.17      0.80   0.05           1 1.40 76  0 27.60\n",
      "SD     0.11      0.45   0.04           0 1.14  0  0  1.14\n",
      "SEM    0.05      0.20   0.02           0 0.51  0  0  0.51\n"
     ]
    }
   ],
   "source": [
    "# Predictivity of the baseline model during 5-fold cross validation\n",
    "metrics <- c(\"MCC\", \"Precision\", \"Recall\", \"Specificity\", \"TP\", \"TN\", \"FP\", \"FN\")\n",
    "perfSum <- matrix(nrow = 5, ncol = length(metrics))\n",
    "colnames(perfSum) <- metrics\n",
    "rownames(perfSum) <- paste(\"Fold\", 1:5, sep=\"_\")\n",
    "\n",
    "# 5-fold cross validation\n",
    "for (i in 1:5){\n",
    "    # Split training & validation sets\n",
    "    valSet <- data[rownames(data) %in% valPairs[unlist(folds[i])], ]\n",
    "    # Baseline probabilities\n",
    "    probability <- combi$stat_prob[match(rownames(valSet), rownames(combi))]\n",
    "    probability[is.na(probability)] <- 0\n",
    "    # Classifiy probability > 0.5 as \"positive\", equivalent to predict(mod_fit, newdata = valProc)\n",
    "    predClass <- factor(ifelse(probability > 0.5, \"positive\", \"negative\"),\n",
    "                        levels = c(\"positive\", \"negative\"))\n",
    "  \n",
    "    # Summarize the model performance on the validation set\n",
    "    perfSum[i,] <- getMetrics(predClass, valSet$Label, positive = \"positive\")\n",
    "}\n",
    "\n",
    "# Mean, SD & SEM of the performance summary\n",
    "perfSum[is.na(perfSum)] <- 0\n",
    "perfSum <- rbind(perfSum, Mean = round(colMeans(perfSum), 2), SD =round(apply(perfSum, 2, sd), 2))\n",
    "perfSum <- rbind(perfSum, SEM =round(perfSum[\"SD\", ]/sqrt(5), 2))\n",
    "cat(\"\\nPerformance summary of the baseline model\\n\")\n",
    "print(perfSum)\n",
    "    \n",
    "# Output  to file\n",
    "write.table(perfSum, paste0(\"Response_to_odorants/Cross_validation/baseline.tsv\"), sep = \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Visualize model performance\n",
    "### Compare the performance of the 5 models with the non-predictive control model (*random*).  \n",
    "We use [Matthews correlation coefficient (MCC)](https://en.wikipedia.org/wiki/Matthews_correlation_coefficient) as the main metric of performance, because of the strong **data imbalance** (which reflects the naturally disproportionate ratio of responsive : non-responsive ORs for a given odorant).  \n",
    "We are also interested in `precision`, which equals the `hit rate` when using the model for virtual screening."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“`fun.y` is deprecated. Use `fun` instead.”\n",
      "Warning message:\n",
      "“`fun.ymin` is deprecated. Use `fun.min` instead.”\n",
      "Warning message:\n",
      "“`fun.ymax` is deprecated. Use `fun.max` instead.”\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdeXxU9b34/08Wwr5YZEc2QaAoSCmKNQVK0QJi3duqbaWlLlC5apfrVqut\nVgu41F4oLtRKFzfQWrWtCG6odYsVFEQjYpVNIewIgYTM74/53vxyISEhhJnkk+fzjz7IOZ+T\nvDlOzcszc2YyEolEAACg7stM9wAAANQMYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJ\nYQcAEIk4w27jxo1btmxJ9xQAACkVZ9i1a9duxIgR6Z4CACCl4gw7AIB6SNgBAERC2AEARELY\nAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC\n2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBE\nIg1hN2PGjNzc3FatWuXm5s6YMaPS9Rs2bPjxj3/cr1+/pk2b9uvX78c//vHGjRtTMCcAQN2S\n6rCbMGHCxIkTCwoKTjnllHXr1k2cOHHSpEn7WL9x48bBgwffeuutbdu2/fa3v92mTZtbb731\nmGOO2bx5c8pmBgCoE1IadgsXLrzjjjtGjRq1ePHiWbNmLVmy5MQTT5w+ffrixYsrOuTmm29e\nvnz59OnTn3322TvvvPO55577zW9+s2zZsttuuy2VkwMA1H4pDbspU6aEECZPnpydnR1CyM7O\nvummmxKJxNSpUys6ZNGiRSGEs846q3RL8s/J7QAAlEpp2M2bN69z5879+/cv3fKFL3yhQ4cO\nTz31VEWHHHPMMSGE+fPnl255+umnS7cDAFAqO2U/adOmTQUFBccff/we27t06fLqq69u3bq1\nefPmex91ySWXPPvss+edd97jjz9+xBFH5Ofnz549e+TIkRdffHHZZUVFRc8//3zpl4lE4mD8\nFQAAarPUhd3WrVtDCK1bt95je3LLli1byg27li1bfuc733nxxRfvv//+5JYGDRqcd955eyze\nunXrCSeccFDmBgCoI1L3VGyDBg1CCBkZGeXPkVn+JL/+9a/Hjx8/ZsyYRYsWffbZZwsXLjzx\nxBO/853v3HrrrWWX5eTknFVGRd8NACBiGSl71rKkpCQnJ+f4448v+5xpCOG44457/fXXd+3a\ntXeNbdiwoVOnTt27d1+0aFGyC0MIu3btOvLII1etWrVmzZoWLVqU+7NycnL69++fl5d3MP4i\nAAC1U+qubGVmZrZt23blypV7bF+1alX79u3Lvcb23nvvFRYWDh8+vLTqQgg5OTnDhg3bvn17\nfn7+wZ0YAKBOSelTlsOHD1++fHnZIFuyZMmKFSuGDh1a7vquXbuGEFavXr3H9jVr1pTuBQAg\nKaVhd+GFF4YQrr/++uSXiUQi+eeJEycmtxQVFa1fv37Tpk3JLzt27DhgwIAnnnhi3rx5pd/k\nySef/Oc//zl48OA2bdqkcngAgFoudXfFhhCGDRs2bty4e++9d/Xq1UOGDHnxxRcXLFgwfvz4\n3Nzc5IIFCxaMHDny6KOPfvPNN5Nb/vSnPx1//PFf+9rXTjzxxB49erz//vtPP/10ixYt/vjH\nP6ZycgCA2i/Vd4/ec889kydPLiwsnDZtWnFx8dSpU2fOnLmP9UcdddR77733gx/8YMWKFbNm\nzVq9evUFF1zw3nvv9enTJ2UzAwDUCam7KzaV3BULANRD3u8NACASwg4AIBLCDgAgEsIOACAS\nwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAg\nEsIOACASwi6dtm/fnpGRsXHjxuptrG8O8HTV83Po7FWbU1dtTh2knrBLj127dk2ZMuXTTz/N\nzMxcv379lClTduzYUfWN6R4/1Q7wdNXzc+jsVZtTV21OHaSLsEuP9evX5+XlDR48uKSk5Jhj\njnnttdcKCgqqvjHd46faAZ6uen4Onb1qc+qqzamDtEnEqEGDBoMGDUr3FJU766yzOnfufPLJ\nJ5eUlFRjY31zgKernp9DZ6/anLpqc+qoJUpKSoqKitI9RYq4Ypc2jz/++BtvvLFw4cJly5bd\nd999+7uxvjnA01XPz6GzV21OXbU5daTd6NGjTz755IcffrhDhw4NGjRo377997///dJXbW7b\ntu1nP/vZEUcc0aRJkx49elx22WVbt25N78A1I91leVDUlSt2a9asSf5v2f8erfrG+uYAT1c9\nP4fOXrU5ddXm1JFeo0aN6tatW6NGjSZMmHD33XefffbZIYRzzz03ufe0007Lzs4+44wzrr/+\n+pNPPrnsrjotI5FIpLsta15OTk7//v3z8vLSPQgAkB6jR49+8sknp0+fPnHixOSW3Nzc999/\n/9NPP928efMhhxxyySWX3HbbbcldY8eOfffdd99///2MjIz0jVwDPBULAMSpUaNGF1xwQemX\nAwYM2L59ewghKysrIyPj6aefXrlyZXLXE088sWzZsrpedUHYAQCx6tq1a3Z2dumXmZn/L3ua\nNWt28803v/vuu126dBkwYMDFF188d+7ckpKSNI1Zk4QdABCnBg0aVLTrsssu+/DDD6dPn96r\nV68HHnhg1KhRw4cP37lzZyrHOxiEHQBQv2zYsOGNN95o0aLFhAkT5syZs3r16v/+7/9+4YUX\nHn300XSPdqCEHQBQvyxZsuSLX/ziPffck/wyJydn+PDhIYSyz9vWUXX+LwAAsF8GDx7cu3fv\nn/70p2+//Xbfvn3ffvvtxx577PDDDx85cmS6RztQrtgBAPVLo0aNnnzyyW9+85tPPfXU1Vdf\nvWDBgrPOOuu5555r2bJlukc7UN7HDgAgEq7YAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEA\nRELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC\n2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBE\nQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEA\nRELYAQBEQtgBAERC2AEARELYAQBEIjvdAwDUenfcER5+uPqHf+c74bvfrblpACok7AAq8/77\nYf786h9+/PE1NwrAvmQkEol0z1DzcnJy+vfvn5eXl+5BgHqgbdvQokVYtizdcwC4YgfAQfPR\nRx898MAD1T68Q4cO3/UsNuwPYQfAwbJp06aXXnqp2of37NlT2MF+EXYAHCx9+/a96667Kto7\na9asl1566Re/+EWHDh3KXdCgQYODNhrESdgBcLDk5OS0b9++or2NGzcOIRx66KH7WAPsF+9j\nBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJ\nYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABCJ7HQPAKTI43e8vmXDjmofPup7A1t3aF6D8wBQ44Qd1Bfz//LWmuUbq334kDFHCDuA\nWk7YQX3xk5mnFO3cXe6uFfkF0y/555dO6fP1iwZXdHinnp87aKMBUDOEHdQXh/U+tKJdJSUl\nIYSWhzbp0b9dCicCoIa5eQKgWnbuDDNnhlGjwoYN4eOPw9ix4c9/DsXF6R4LqNeEHcD++/jj\ncOyx4fzzw9y5YffuUFQU/v738J3vhC9/Oaxdm+7h6oYNGzasW7cuhLB8+fJiQQw1RNgB7Ked\nO8Po0WHRonJ2vfJKOPnksLv81zKStGbNml/84hff+973Fi1aFEK45ZZbvve97z366KOJRCLd\no0GdJ+wA9tOdd4Z33qlw72uvhb/8JYXT1DHLly+/7LLL3njjjbIZt3nz5nvuuefWW2/VdnCA\nhB3Afnr44QNdUF/t3r375ptv3r59e7l7n3/++fnz56d4JIiMsAPYT8uWVbLg/fdTMkfd88Yb\nb6xcuXIfCx599NGUDQNREnYA+6nSpws9n1iBxYsX73vBihUrtmzZkpphIErCDmA/9exZyYJe\nvVIyR91TlWjbvHlzCiaBWAk7gP105pkHuqC+atasWY2sASoi7AD20wUXhL59K9w7eHA499wU\nTlOX9N3HeQshhNC+fftDDjkkNcNAlNIQdjNmzMjNzW3VqlVubu6MGTOqcsgLL7wwcuTIli1b\nduzY8Zvf/OYHH3xwsIcEqFCjRuGf/wxHHVXOrmOOCY8/HrKyUj5T3TB48OC2bdvuY8FJJ52U\nsmEgSqkOuwkTJkycOLGgoOCUU05Zt27dxIkTJ02atO9DHnzwwZEjRy5dunTs2LHHHHPMww8/\nfOyxx3788cepGRigHF27htdfD3fcEUaODJmZITs7jB4d7r03vPhiaOfzdiuUk5Nz2WWX5eTk\nlLv3C1/4wtixY1M8EkQmI5XvBrlw4cKBAweOGjXq8ccfz87OLi4uPumkk+bNm/fWW28deeSR\n5R7y8ccfH3744YMGDZo7d27Lli1DCDNnzjz//PPPO++8e++9t6IflJOT079//7y8vIP0F4HI\nLFu45qqT/jJ6/Be+98sR6Z6lDmrbNrRoUfl7oPC/li1bNn369LLPveTk5IwdO/bcc89t0KBB\nGgeDCGSn8odNmTIlhDB58uTs7OwQQnZ29k033fTUU09NnTp11qxZ5R5y++23FxcX/+Y3v0lW\nXQhh/Pjx27Zta9SoUcrGBqAG9ezZ87bbblu+fPmdd965dOnS888/f/jw4c2bN0/3XBCDlF6x\na9OmTaNGjVasWFF2Y8eOHROJxJo1a8o9pFOnTllZWfv7xKsrdlBFH72z7rEZr73x9PLtm3dm\nZmX0Htxp5LkDck/rm5GR7snqEFfsquu3v/3t/Pnzp0+ffthhh6V7FohE6l5jt2nTpoKCgq5d\nu+6xvUuXLp988snWrVv3PmTbtm2rV6/u1q3bokWLvv71r7dr165Lly5nnXXWsr3+BVpcXPxG\nGT5tEKpi/p/fumL0n154ZOn2zTtDCCW7E0tfWfk/k/4+5Xt/LdrlY+wB6p7UhV0y3Vq3br3H\n9uSWct+1ctOmTSGE1atX5+bm/uc//xk7dmy/fv0eeeSRAQMG7HE1bsuWLV8so7i4+GD9NSAW\ni1/6+O4r5+0uLtl71xvzPvjDNU+nfiQADlDqXmOXfElsRgVP8GRmlpOYRUVFIYQPPvjgiiuu\nuPHGG5PHzp8//8QTT7zgggv+/e9/l/3mI0eOLP3ymWeeqdnhIT4PTH4xUVLhte2n73v7lInH\ntOvaKpUjAXCAUhd2bdu2zcrK2rhx4x7bN2zYkJWV1a68Nwho0qRJCKF169Y33HBDaRGOHDny\nhBNOeOqpp9auXVv6fkjNmzefN29e6YEV3UsPJG0u2P7+v1fvY0GiJPHv+ctHj/9CykYC4MCl\n7qnYzMzMtm3brly5co/tq1atat++fblX7JI3W3Tv3j3r/77bZ48ePUIIe38roIoKVm2p9JWo\na1f4yE6AOialb1A8fPjw5cuX5+fnl25ZsmTJihUrhg4dWu76zMzM4cOH5+fnFxYWlt2+dOnS\nzMzM3r17H9xxIV7ZOZV/NEJ2Ax+fAFDHpDTsLrzwwhDC9ddfn/wykUgk/zxx4sTklqKiovXr\n1yfvmUi67LLLtmzZ8pOf/KSk5P+9xPuhhx56/vnnx4wZ07Rp01QODzFp16VVg8rarvMRe97q\nBEAtl9I3KB42bNi4cePuvffe1atXDxky5MUXX1ywYMH48eNzc3OTCxYsWDBy5Mijjz76zTff\nTG458cQTx40bN3369AULFhx33HEffvjhvHnzOnToUMUPmQXK1ahpgy+e2PPlJ96raEHjZjmD\nTjg8lSMBcOBS/Vmx99xzz+TJkwsLC6dNm1ZcXDx16tSZM2fu+5A//OEPN998c8uWLe+///5V\nq1ZdfPHFS5Ys6dy5c2oGhlidc9WXm7as8BNczrlqaLNWPt8FoI5J6SdPpIxPnoCq+GDhJ7de\n9Pi6/3uTRFZ25jlXDT35wi+ma6q6xydPVJdPnoAal9KnYoFa5fCj29/23Pde/OvSlx5d+vaL\nHx/aucVXvnnk8G8c2aZzi3SPBkB1CDuo13IaZY84+6gufQ99+6S/DP5az7N+9KV0TwRA9aX6\nNXYAABwkwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4A\nIBLCDgAgEsIOACAS2ekeoF5a/3r49JnqH96qf+g4uuamAQAiIezSYd2LYeEV1T/88O8LO6rh\n6b+8tW1TYbm71q/eGkJY/tanf5v+WkWHDz2r3yFtmx6s4QCoCcIuHTqdFJp0rnDvwsvDZyvC\n8fdVuKBZ94MxFNF77I7X1yzfuI8F772+6r3XV1W098jjuwg7gFpO2KVD8yNC8yMq3PvOr8P2\nVaHLWSkciHrhe78cseOzXdU+vF23VjU4DPXE7t27d+zYUdHeoqKiEML27du3bdtW7oLMzMwm\nTZocrOEgRsIO6oujv+Jab3X9+Mfh1lv3tWDdupCRUeHea68N111X0zPVDYsXL77mmmv2vean\nP/1pRbt69ux5677PPPB/CTuAynTuHAYNqv7hHTrU3Ch1TPPmzQcMGFDtwzt16lSDw0B9kJFI\nJNI9Q83Lycnp379/Xl5eugeplicHhU2Lw7d2pnsOAKCO8T52AACREHYAAJEQdgAAkRB2AACR\nEHYAAJEQdgAAkfA+dtQlK4uKvvef/1T78FZZWbN79Ki5cQCgdhF21CXbS0rmb91a7cPbZnvA\nAxAzv+eoSw7PydlQ8bvY//rTT6d88snDPXp8pXnzchdU/JFPABADYUddkpWRcUhWVkV7G2Vk\nhBCaZWbuYw0ARMzNEwAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYA\nAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2\nAACRyE73AFAD3iksvKugYM6mTSGEn65albd9+/mHHtom28MbgPrFbz7qvF998sm1a9bsTiSS\nX761Y8dbO3bc/Omnf+7WbUzLlumdDQBSyVOx1G3T1q372erVpVVXauPu3Wd8+OHr27enZSoA\nSAthV5usfy28cWnY+n4oKQqvXRDWPBnCnr1CWVt277569eqK9haWlPxk5cpUzgMA6eWp2Nph\nd2F47cLw4R///y3L7g7L7g7tRoTcB0LDNumbrFZ7csuWLbt372PBC9u2rSkq6tCgQcpGAoA0\ncsWudnjle/+n6kp9+kx4dkwoKUr5QHXD+zt37ntBIoT8ytYAQDSEXS3w6bPhowcq3LshLyy7\nM4XT1CV7v7RubyVVWAMAcRB2tcB/7jvQBfVVj4YNK11zeBXWAEAchF0tsGVpJQs2v5OSOeqe\nUS1aNM7c12N4UJMmXXJyUjYPAKSXsKsFSnYd6IL66tDs7Cvatatob1ZGxpROnVI5DwCkl7Cr\nBZp2rWxBt1SMUTf9rEOHiW3KuWu4UWbmzC5dRjRvnvqRACBdhF0t0OnkShZ0rmxBPZYZwvTD\nDnu2V69vHHJIi6ysEMJhOTk/bNPm7b59x7Vune7pACClhF0t0O2c0OqoCvc2bB36/iSF09RJ\nw5s3f7B790vbtg0hzOzSZdphh/V0zwQA9Y+wqwUyssOwx0LzI8rZ1bBNGPa4NygGAKpC2NUO\nTbuF0W+EATeGVkeGjBBCCM0OD31/Ek56Oxx6XJpnAwDqCB8pVmtkNwv9rgz9rgxPDgqbFoev\nL0v3QABAHeOKHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSy0z0A7IcPdu4c\n/O67Fe0tLCnJCOHM5cuzMzLKXXBodnZ+v34HbToASDNhR13SICOjR8OG1T78c9ke8ADEzO85\n6pIuOTl5ffqkewoAqKW8xg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAg\nEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLZ6R6gXnr3tvDvH1Wy\n5r6MCncd/v1w7O9rdiIAIALCLh2adgntR1b/8Bafr7lRAIB4CLt0OOyMcNgZ6R4CAIiN19gB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC\n2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEoo6F3bZt22bNmrVy5cp0\nDwIAUOukIexmzJiRm5vbqlWr3NzcGTNm7NexkyZNGjdu3KJFiw7SbAAAdVeqw27ChAkTJ04s\nKCg45ZRT1q1bN3HixEmTJlXx2Dlz5tx7770HczoAgDospWG3cOHCO+64Y9SoUYsXL541a9aS\nJUtOPPHE6dOnL168uNJjV61adeGFFzZr1iwFcwIA1EUpDbspU6aEECZPnpydnR1CyM7Ovumm\nmxKJxNSpU/d9YCKR+O53v9uyZcv/+q//SsWgAAB10H6E3bp168p++cQTT7z00ku7du2q+neY\nN29e586d+/fvX7rlC1/4QocOHZ566ql9H3jLLbc899xzf/zjH1u0aFH1HwcAUK9UKeymT5/e\nt2/f73//+2U3zpo1Kzc3t3Xr1r/97W8TiUSl32TTpk0FBQVdu3bdY3uXLl0++eSTrVu3VnTg\nwoULr7766ssvvzw3N7eiNYlEYmMZlQ4DABCf7EpXnHvuuffdd192dvZJJ520x/aMjIzHH3/8\nkksuefnll++///59f59kurVu3XqP7cktW7Zsad68+d5H7dix49xzz/385z9/3XXX7eObb9y4\nce/vDABQr1QSdn/729/uu+++4447bs6cOR07diy769RTTz311FOXL18+YsSIBx544Lvf/e7o\n0aP38a0aNGgQQsjIyCh3b2Zm+dcOf/rTny5fvjwvLy8nJ2cf3zwzM7NHjx6lX3744Yf7WAwA\nEKVKwm7atGnZ2dmzZs3ao+pK9ejRY/78+b169frd736377Br27ZtVlbW3s+TbtiwISsrq127\ndnsf8vTTT0+fPv22227r16/fvuds1arVBx98UPrlvisQACBKlbzGLj8/f8CAAb169drHmp49\nex5xxBH5+fmV/KTMzLZt2+79oRGrVq1q3759uVfsFi5cGEK47LLLMv7XFVdcEUIYO3ZsRkbG\n73//+33/RACAeqWSK3YFBQWDBg2q9Lv069dv7ty5lS4bPnz4/fffn5+ff8QRRyS3LFmyZMWK\nFWeffXa56wcMGHDRRReV3fLmm2+++uqro0eP7tq1a58+fSr9iQAA9UclYde7d+/333+/0u/y\nzjvvdO/evdJlF1544f3333/99df/6U9/CiEkEonrr78+hDBx4sTkgqKioi1btmRlZbVq1SqE\nMHLkyJEjR5b9DpMnT3711Vd/+MMf7nEnBwAAlTwVe9RRR73zzjsff/zxPta8++67+fn5ffv2\nrfSHDRs2bNy4cX/+85+/+tWvXn311cOHD3/wwQfHjx9f+j4mCxYsOPTQQ7/yla9U/S8AAEBS\nJWF30UUXZWRknHbaadu3by93wdatW08//fREIjFhwoSq/Lx77rln8uTJhYWF06ZNKy4unjp1\n6syZM/d7agAA9pJR6XsL/+xnP/vVr37VtWvXK6+88qyzzvrc5z6X3F5QUPDggw/+6le/WrNm\nzfnnn3/XXXcd/GmrKicnp3///nl5eekeBAAgdSoPu5KSkrvvvvuqq67asGFDCKFFixYdOnRY\ns2bNli1bQghNmza97rrrLr300uTHv9YSwg4AqIcqD7ukgoKCqVOnzp07Nz8/f8eOHTk5Od27\nd//qV796+eWXd+nS5WBPub+EHQBQD1U17MrauXNnw4YND8Y0NUXYAQD1UCU3T5S1bt265B+S\nVffEE0+89NJLu3btOihzAQCwn6oUdtOnT+/bt+/3v//9shtnzZqVm5vbunXr3/72t9W47AcA\nQM2qPOzOPffciy++eNmyZb17995j+1lnnVVcXHzJJZecc845B21CAACqpJLX2P3tb3879dRT\njzvuuDlz5nTs2HHvBcuXLx8xYsRHH330j3/8Y/To0Qdtzv3jNXYAQD1UyRW7adOmZWdnz5o1\nq9yqCyH06NFj/vz5IYTf/e53NT8dAABVVknY5efnDxgwoFevXvtY07NnzyOOOCI/P79GBwMA\nYP9UEnYFBQVVeZu6fv36rVy5soZGAgCgOioJu969e7///vuVfpd33nmne/fuNTQSAADVUUnY\nHXXUUe+8887HH3+8jzXvvvtufn5+3759a3QwAAD2TyVhd9FFF2VkZJx22mnbt28vd8HWrVtP\nP/30RCIxYcKEgzAeAABVVUnYHXfccVdcccW///3vz3/+83feeeeGDRtKdxUUFEyfPr13795L\nly49//zzR4wYcZBHBQBgXyr/rNiSkpK77777qquuSlZdixYtOnTosGbNmi1btoQQmjZtet11\n11166aXZ2dmpmLdqvI8dAFAPVR52SQUFBVOnTp07d25+fv6OHTtycnK6d+/+1UzuiUcAACAA\nSURBVK9+9fLLL6/KbbMpJuwAgHqoqmFX1s6dOxs2bHgwpqkpwg4AqIcq/6zYvdXyqgMAqJ8q\neWHcq6++WvXvdeyxxx7YMAAAVF8lYTdkyJCqf69qPKsLAEBNqdKtrC1atBgzZkznzp0P9jQA\nAFRbJWF31VVXzZkzJz8//8EHHxwyZMiZZ555+umnd+vWLSWzAQCwH6p0V+zixYsffvjhhx9+\n+O233w4hDBo06MwzzzzjjDN69ep18CesDnfFAgD10P693Ul+fn6y8N54440QwlFHHZUsvH79\n+h20CatD2AEA9VB13scuhPDRRx8lC+/ll19OJBK9e/c+88wzb7jhhhqfr3qEHQBQD1Uz7Eqt\nWrXqmmuu+cMf/hBq012xwg4AqIeq/wGvr7/++pw5c2bPnv3hhx9mZGTs1xujAABQ4/Yv7BKJ\nxGuvvTZ79uw5c+Z89NFHmZmZX/rSly699NLTTz/dm6EAAKRXlcIukUi8/PLLyZ5buXJlZmbm\n0KFDf/rTn55++ukdOnQ42CMCAFAVlYTdCy+8MGfOnDlz5qxevTorK2v48OFXX331aaed1q5d\nu9TMBwBAFVVy80RGRkYIoXnz5ieffPJpp53Wpk2bfSweNmxYDU9XXW6eAADqoSqFXRW5KxYA\nII0qeSr2uuuuS8kYAAAcqAN9H7vayRU7AKAeykz3AAAA1AxhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEH\nABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlh\nBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJ\nYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEH\nABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlh\nBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEInsdA8AAFCTPv30\n088++6zah7dp06Z58+Y1OE8qCTsAICqTJk2aPXt2tQ+/++67f/CDH9TgPKkk7ACAqIwcOfKQ\nQw4pd9fu3bt///vfd+zYcezYsRUd3qdPn4M22kGXkUgk0j1DzcvJyenfv39eXl66BwEAapHC\nwsLGjRsPHz782WefTfcsB4WbJwAAIpGGsJsxY0Zubm6rVq1yc3NnzJhR6fqdO3deffXVQ4cO\nbdmy5eGHH37OOed88MEHKZgTAKBuSXXYTZgwYeLEiQUFBaeccsq6desmTpw4adKkfazfvHnz\nCSeccOONN65fv/4b3/hGz549H3jggaOOOmrhwoUpmxkAoE5IadgtXLjwjjvuGDVq1OLFi2fN\nmrVkyZITTzxx+vTpixcvruiQyZMnv/DCCz/84Q+XLFly9913z50794knnti5c+e4ceNSODgA\nQB2Q0rCbMmVKCGHy5MnZ2dkhhOzs7JtuuimRSEydOrWiQ/761782b978lltuKd0yZsyYESNG\nLFq0aO3atSmYGQCgrkhp2M2bN69z5879+/cv3fKFL3yhQ4cOTz31VEWHZGZmDhs2rGHDhmU3\n5uTkhBA2btx48EYFAKhzUvc+dps2bSooKDj++OP32N6lS5dXX31169at5b7L85IlS/bYsm7d\numeeeaZdu3aHH374wZoVAKAOSt0Vu61bt4YQWrduvcf25JYtW7ZU5Zvk5+d/6UtfKiws/PWv\nf518Pjdpw4YNGWUUFRXV3OAAAHVD6q7YNWjQIISQkZFR7t7MzEoS87PPPpsyZcrUqVMTicS0\nadP2uHkiIyOj7HtMb9q06UDHBQCoa1IXdm3bts3Kytr7hXEbNmzIyspq167dPo795z//edFF\nF3388cdjx469+eabe/fuvceCQw45ZMOGDaVfJl+EBwBQr6TuqdjMzMy2bduuXLlyj+2rVq1q\n3779Pq7YXXvttWPGjGnevPnzzz//+OOP7111AACEFN8VO3z48OXLl+fn55duWbJkyYoVK4YO\nHVrRIbNmzfrlL3/5rW9969///vc+lgEAkNKwu/DCC0MI119/ffLLRCKR/PPEiROTW4qKitav\nX1/6CrlEInHTTTd16tTpD3/4g2dXAQD2LXWvsQshDBs2bNy4cffee+/q1auHDBny4osvLliw\nYPz48bm5uckFCxYsGDly5NFHH/3mm2+GED766KP33nuvTZs2p5122t7f7U9/+tOhhx6ayvkB\nAGqzlIZdCOGee+7p27fv3/72t2nTph155JFTp079yU9+UtHi5cuXhxDWrVv35JNP7r13586d\nB3FQAIC6JiORSKR7hpqXk5PTv3//vLy8dA8CANQihYWFjRs3Hj58+LPPPpvuWQ6KlL7GDgCA\ng0fYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEA8UskEnPnzr3kkktCCG+//fY111zz7rvvpnuompeRSCTSPUPNy8nJ6d+/f15eXroH\nAQDSb82aNWefffbzzz9fdmNmZuall146ZcqUrKysdA1W47LTPQAAwEG0Y8eOr33ta2+//fYe\n20tKSm699dbi4uLbb789LYMdDJ6KBQBidvvtt+9ddaWmTZu2cOHCVM5zUAk7ACBmf/7zn/ex\nt6Sk5L777kvZMAebsAMAorV79+5Kb5JYsmRJaoZJAWEHAESruLh49+7d+15TWFiYmmFSQNgB\nANFq2LBhu3bt9r2ma9euqRkmBYQdABCzsWPHHuCCOkTYAQAxu+qqq5o0aVLR3sGDB5966qmp\nnOegEnYAQMx69Ogxe/bsZs2a7b2rb9++f/3rXzMz48mheP4mAADlGjNmzKJFi84///w2bdok\nt/Tt2/fGG2/My8vr1KlTemerWT5SDACoLwoLCxs3bjx06NA9Pl4sGq7YAQD1S0zPve4h2r8Y\nAEB9I+wAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh\n7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAi\nIewAACIh7AAAIiHsAAAiIewAACIh7AAAIpGd7gEAAGrS22+//emnn5a7a9euXSGEjRs3zp8/\nv6LD+/bt26lTp4M13EGWkUgk0j1DzcvJyenfv39eXl66BwEAUu0b3/jG7Nmzq3343Xff/YMf\n/KAG50klV+wAgKice+65gwYNqvbhxxxzTA0Ok2Ku2AEARMLNEwAAkRB2AACREHYAAJEQdgAA\nkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACRyE73\nAABAOd57772SkpJqH96rV6/sbL/l6x3/yAGgNnrppZeKi4urfXjXrl2FXT3kHzkA1EbHH398\nRVfsCgoKli5d2r17986dO1d0eE5OzkEbjdpL2AFAbdS7d++Kdn344YdLly5t27Zt3759UzkS\ntZ+bJwAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHs\nAAAi4bNiAaDO2LVrV35+/rJly0II7733XiKR6N27d+PGjdM9F7VFRiKRSPcMNS8nJ6d///55\neXnpHgQAaszKlSufeeaZwsLCshsbNGjw5S9/uWfPnumailrFU7EAUAesXbt27ty5e1RdCKGo\nqOjZZ5/96KOP0jIVtY2wA4A64F//+tfu3bvL3ZVIJF566aWSkpIUj0QtJOwAoLbbsmXL2rVr\n97Fg27Ztn3zyScrmodYSdgBQ223atKnSNRs3bkzBJNRywg4AaruqPM3qqViCsAOA2q9Zs2aV\nrmnevHkKJqGWE3YAUNu1bt16322XnZ3dsWPHlM1DrSXsAKC2y8jIGDx48D4WDBw4MCcnJ2Xz\nUGsJOwCoA3r16jVo0KByd/Xp0+foo49O8TzUTj5SDADqhkGDBnXq1Ontt99esWJFcXFxZmZm\np06d+vXr16VLl3SPRm0h7ACgzmjfvn379u0//PDDefPmDR48eMCAAemeiNrFU7EAAJEQdgAA\nkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkfBZsQBQG23d\nujWRSJS7a/v27SGEwsLCLVu2VHR48+bNMzIyDtZw1FbCDgBqo9mzZxcXF+9jwaJFixYtWlTR\n3m9/+9tNmjQ5CHNRqwk7AKiNunXrVlJSUu3Ds7P9iq+P/FMHgNpoxIgR6R6BusfNEwAAkRB2\nAACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQ\ndgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACR\nEHYAAJEQdgAAkRB2AACRSEPYzZgxIzc3t1WrVrm5uTNmzDhIhwAA1DepDrsJEyZMnDixoKDg\nlFNOWbdu3cSJEydNmlTjhwAA1EMZiUQiZT9s4cKFAwcOHDVq1OOPP56dnV1cXHzSSSfNmzfv\nrbfeOvLII2vqkBBCTk5O//798/LyDtpfBQCg1knpFbspU6aEECZPnpydnR1CyM7OvummmxKJ\nxNSpU2vwEACA+imlV+zatGnTqFGjFStWlN3YsWPHRCKxZs2amjokuGIHANRLqbtit2nTpoKC\ngq5du+6xvUuXLp988snWrVsP5JCNGzd+rozi4uIanx8AoJbLTtlPSnZY69at99ie3LJly5bm\nzZtX+5BEIrFx48aDMDUAQJ2RurBr0KBBCCEjI6PcvZmZ5Vw7rPohn/vc58o+p5yTk3MgowIA\n1EWpeyq2bdu2WVlZe19X27BhQ1ZWVrt27WrkEACAeit1YZeZmdm2bduVK1fusX3VqlXt27cv\n94pdNQ4BAKi3UtpGw4cPX758eX5+fumWJUuWrFixYujQoTV4CABA/ZTSsLvwwgtDCNdff33y\ny0QikfzzxIkTk1uKiorWr1+/adOmqh8CAEBS6m6eCCEMGzZs3Lhx99577+rVq4cMGfLiiy8u\nWLBg/Pjxubm5yQULFiwYOXLk0Ucf/eabb1bxEAAAklL9MrV77rln8uTJhYWF06ZNKy4unjp1\n6syZM2v8EACAeiilnzyRMj55AgCoh9xYCgAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlh\nBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJ\nYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEH\nABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlh\nBwAQiYxEIpHuGWpeTk5O69atzzvvvHQPAgBQk66//voGDRpUuDsRo5kzZ6bwDAMApMhnn322\njwSK84pdUVHRW2+9le4pqmns2LGffPLJ888/37Rp03TPUsdMnjx59uzZN9xww6hRo9I9Sx3z\n0ksvXXLJJaNGjbrhhhvSPUvdM3jw4KZNmz733HPpHqTu+dGPfrRgwYK777574MCB6Z6ljnnk\nkUduvPHG8847b9KkSemepY5Zu3btmDFjDj/88AcffDDds1TTwIEDMzMrfClddipHSZkGDRoM\nGjQo3VNUU05OTgjh6KOPbtGiRbpnqWPatGkTQujevXvd/aefLp9++mkI4XOf+5xTVz1ZWVlO\nXTW0bNkyhNC7d29nb3/l5eWFENq3b+/U7a+VK1eGEBo3bhzrqXPzBABAJIQdAEAk4nwqtk77\n3e9+t3379saNG6d7kLpn/Pjxw4cPP/bYY9M9SN0zcODAhx56qEuXLukepE568MEH93WHGhW7\n/PLLv/Od7/Tt2zfdg9Q9J5xwwkMPPdSnT590D1L3tG7d+qGHHkq+DCBKcd48AQBQD3kqFgAg\nEsIOACASwg4AIBLCrjbavHlzukegPvLAA6jrhF2tM2fOnO7du7/yyivpHoT6xQMPIALCrnaZ\nM2fO2WefvXHjxjvvvDPds1CPeOCRYrt37073CBAnYVeLJH+5NmrUaMSIEYsWLUr3OHVJfn7+\n17/+9e7du3fo0OGb3/zmq6++mu6J6hIPvGrzwKuenTt3Dhw48OGHH073IBAhYVdblP5y/ec/\n/zl+/Ph33nmnuLg43UPVDa+//nr//v3nz5/ft2/fdu3aPfTQQ8cdd9x//dd/7dq1K92j1QEe\neNXmgVdtDRs2POSQQ771rW9puwPx2WefzZ49++GHH96xY0e6Z6mTCgsLP/nkk5KSknQPUtMS\n1AKzZ8/Ozs5u1qzZCy+8kEgk3n333RDCwoUL0z1XHVBSUjJw4MBDDz30P//5T3LLU0891atX\nrxDCiBEjtmzZkt7xajkPvGrzwDtAkydPDiFkZ2fPmTMn3bPUSR9++GHv3r2Tv8f79eu3bNmy\ndE9Ul3z66afnnHNOw4YNQwgtW7a87LLLYvr/rLBLvz1+uSYSiZKSkubNm990003pHaxOWLt2\nbQjhZz/7WdmN69evHzp0aAjhhBNOKC4uTtdstZwH3oHwwDtAF1xwQTJKtF01lJSUDBs27JBD\nDrnyyiu//e1vZ2RkdO7cWdtVUUFBwWGHHZaVlXXGGWdMmjQp2cdHHXXUqlWr0j1azRB2abZ+\n/foWLVqU/eWadOKJJ37pS19K11R1SPIi0xVXXLHH9p07d37lK18JIfzoRz9Ky2C1nAfeAfLA\nOxCLFy9u1KhRnz59Zs+e3axZM223vxYtWpSZmfnaa68lv3zwwQezs7O1XRWdffbZDRo0eP75\n55Nf7tq1a9CgQSGEmTNnpnewmiLs0u/pp5/e45drIpG4/fbbMzMz165dm5aR6pCdO3c2adLk\nsMMOKyws3GPXxo0bkx8u/swzz6RltlrOA+9AeOAdiJ///Od9+vRZs2ZNIpFYsGCBttsvO3bs\nePrpp7t161Z240MPPaTtqmL79u0ZGRkTJ04s3XLttdeGEG677bZEIvHUU0+lb7QaI+zSoCrP\n0SxbtiyE8Otf/zoF89QtRUVFr7zyyuzZsxcvXpzcMmnSpBDCz3/+870X5+XlZWRknHbaaamd\nsfba++ztwQOvIkVFRS+//PKCBQs2btyY3OKBV3VFRUVlv9y9e/f69etLv9R2VfT+++8nry2F\nEJo0abJz586ye7VdVSxZsiSEcPvttye/LFt1v/vd7xo0aLBy5cq0DlgDhF2qFRYWHnXUUVX5\n91f//v07deq0a9euFExVV7z88svdunUrvfXn1FNPXbdu3fr16zt27JiRkXHfffftfcjYsWOz\nsrJ27NiR+mlrm3LP3t7LPPD29swzz3Tu3Dl53po1a5Z8EscDrypeeeWV4cOHZ2dnN2zY8Kyz\nzpo7d265y7RdpZYuXdq6detGjRoNHjy4UaNGIYQrr7xyjzXarlJbt25t2LDhOeeck/i/VZdI\nJB566KEQwt///ve0DlgDhF0aDB06tCr//kq+Veyf/vSn1ExV+z399NMNGjRo3779DTfcMH36\n9CFDhoQQvvrVryYSiRdffLFx48YNGzacPXv2Hkf95Cc/CSF4Jfs+zt4ePPD2kJeX17Bhw8MO\nO+zWW2+95ZZbrrnmmpKSkuQuD7x9e+6553Jyclq2bPmtb33rrLPOatGiRQjhG9/4xrZt2/Ze\nrO327eKLL27atOlbb72VSCSWLVvWp0+fEMJvfvObPZYl2+7WW29Nx4y10Y4dOx555JG77rpr\nxYoVyS2jR4/OzMz89re/XbbqEv8bdhX9t0cdIuzSoIr3+X/22WeHHHJIt27dPvvss5TNVmvt\n2LGjU6dOhx566Keffprcsnv37oEDB4YQ3njjjUQi8Y9//KNp06YZGRm//vWvd+/eXXZN586d\n0zZ37VDp2SvLA28PJ554YrNmzT7++ONy93rgVWTbtm3du3fv0aNH6S/Ujz/++MgjjwwhDBky\npPQZ7bK0XbmWLl16ySWXnHvuuT/4wQ9KN65ZsyZ5O+febbdkyZLUDlh7vfvuu/369Utea2/R\nosXjjz+eSCTy8/ObN28eQrjssstKVxYXF3/5y19u2bJlBFfZhV0aVP0+/5///OflXm+vhx57\n7LEQwh7vxHHfffeFEGbNmpX8ctGiRV27dg0hHHnkkXfdddejjz46evTosgvqraqcvbI88EoV\nFRU1bNiw7C/UtWvXXnrppUOGDBkxYsSNN95YWFjogVeuv//97yGEBx54oOzG1157Lflvv+OP\nP3779u17H5Vsu1/+8pepGrMO+NWvfpX8fVH2Jf+JfbYdiUTio48+at269SGHHPLLX/7yF7/4\nRcuWLXNych577LFEIjFnzpycnJymTZtOnTr1X//612OPPfbFL34xhHD//fene+oaIOxSbb/u\n89+2bVuHDh0aNGgQzfvrVNtdd90VQrjjjjvKbnzwwQdDCH/5y19Kt6xdu/aHP/xh8r/GQgiN\nGzeeNm1ayoetdap49kp54JVKfozEBRdckPzy9ddfb9u2bQihXbt22dnZyZhbt26dB97epk6d\nuvcl4d27d2dnZ/fs2TOE8P3vf7/cAz/66KOUDFiXXH755clHXUFBQdnt2m4frr322uzs7NJH\n4NKlS9u3b1/adq+99lr37t1LX3PcsmXLP/7xj2mdt8YIu1Tb3/v877nnnmuvvTZ189VWr7zy\nSghh7NixZTfefPPNIYS9n3fYsmXLo48++sgjj5R7c0A9tF9nL8kDr1TPnj3btGlTWFi4c+fO\nz3/+8126dHnllVcSicTmzZvHjRsXQjjzzDOTKz3wynrkkUdCCLfcckvZjckPI87Ly0s+QVb6\nXmJUKtl2xx577ObNm8tuL227l19+OV2zpdfOnTtvuOGGsm88lHzy+oc//OHQoUPLrnznnXfK\ntl1RUdFzzz03efLk+++/P/lLOQ7CLtXc519tubm5IYT58+eXbhkzZkynTp1KX8meSCQ++eST\ncp/fwdmrtuuuuy6EcM011zzzzDMhhH/961+lu0pKSpJvR/zhhx+mb8Baav369a1bt27WrFnp\nGdu0adOQIUM6dOiwffv2vLy85I0U6R2yltvj5vR9tN3//M//pHa0WuS2224LIXz9618v3ZJ8\n8rpVq1bHH3/8Hov3aLsoCbv003ZVtHLlyiuvvLI0RHbt2tWkSZMJEyaULlizZk2fPn1GjBix\nx5tmkXD2DsDmzZu7du2akZFx9tln730/xL333htCePjhh9MyW62y9y3Ac+bMycjIyM7OPuOM\nMyZOnNi5c+eMjIxHH300ufe4447r2rVrqqesC7Zs2XLRRRd17NgxMzPzuOOOK/siiorarj4r\nLi4eN27c3/72t7IbkycqKyvr3Xff3WN99G0n7GoFbVcNyf/iL/2FmuySEIInEKvC2dsvr7zy\nSpMmTZJvYrdHvsycOTOE8NJLL6VrtlqionfonD9/fq9evZIvY+rZs2fZa8annnpq27ZtUztm\nHbBmzZpu3bplZGQMHTp0zJgxjRs3Tl6OKr2PWNsl7dy587rrriv7JMPSpUu/+93vlj4nmzxR\nRx111N4nKtl2119/ferGTSFhV1tou/2VfLu1/Pz8hC7Zf87e/vrHP/7RsGHDEELZXwa7du0a\nOHDgoYce6q1hEvt8h841a9bs8RqmTZs2tWvX7mtf+1qqpquNdu7cufdtmGPHjs3JyXn66aeT\nX27cuPGMM84IIQwdOrT0PypK227vT7SrP37/+9+HEEaMGFHadj/+8Y9DCKNGjdqj7cqN4Ihf\nCCvsUqGwsPAXv/jF/9fe3UdFVed/AP/cYYYhYQQ0DHIEVMqTSLr4APi0VLuKbm4E2DmR5rHa\nlExhNwsXahEt6MEH8gk18ykNSx4StULtSOCRlNjawAADERBotNwSFBgY7u+P79l75ncHh/GC\n3GHm/frDw3zvV/zM5wzOm7n3fr+zZs0KCgpav3797dYsxX3+psy0btu2bURUVlaGXHI76J5k\n3bbu1KlTbm5uRPTMM8989tlnBw4cCAwMJCLT1YntkyUrdBYWFl69erWqqmrWrFkcx508ebI/\nK7Q2q1atIqLVq1cLI2xPP+P11ZjnnnuOXegpjMTHx+NnNjY21jjbGQyGxYsXW57tbBWC3V1X\nV1fHdgT38fFxcXFxdHQsKCi43WTc52/MfOuOHDlCREuWLEEu6Ra6J5mZ1l2+fPmZZ55hZ8fY\nmdn09HR5q7UePa7Q+cknn3Acx+aoVCphv0679dtvv02ZMsU4233xxRdEtGPHDtHM1tbWUaNG\nDRkyxM73MjGFbGcKwe6uCwsLGzRoEPtcvaWlpbi4WO6KBgzzrdPr9WxFWFiECAAAE5lJREFU\nMeSSbqF7kvX4M/v777/n5uYeOXLk+vXrchRojSxZobO1tXXHjh0xMTGJiYlsaywQZbvKykoi\nWrRokenMxMREIjp//nx/l2j17ijb2cPJawS7u+uHH34gouTkZNND33333Y0bN/q/pIHCktZl\nZmYil3QL3ZMMP7OWaG9vF+2xdqcrdIJAlO3Gjh3r5ORkutff2rVrhetiQcTCbGcn/90h2N1d\n7IRXTk6OaFyn07m7u8fFxclS1YBgYetOnz7d35UNBOieZPiZ7VF7e/tf//pXrVZbVVUlDGKF\nTmkMBkNOTk5qaqpKpWLZ7uzZsyqVysfHp7y8XJjGFsf28vLCqdjbsSTb2QkEu7uruLiYiFat\nWmV6iN1M1/8lDRRoXW+ge5KhdT1KTk5mJ/FF2U4E2a5Hly5dCgoKIiIPD48hQ4awrq5evTo9\nPV2hULi4uKSkpPznP//Jz8//4x//SETHjh2Tu2Qr0tra2tTUZDAYhBEz2c6uAjGC3V0hLBfe\n1dWl1Wrd3d1NF6Z//vnnOY5rbW3t7+KsHuseWicBXniSoXWWq6iocHBwYAv7IdtJ1tXVNW3a\nNLVazZaT7OzszM3NHT58OMt2J06c8PDwoP/RaDS7d++Wu2RrodPpoqOj2fJDrq6uf//734Vr\nJLrNdnZyBlaAYNeXul0unG2YOH78eJ1OJ8xsa2t78MEHJ0yYIF+xVse0e2idhfDCkwytk2bR\nokVqtTomJgbZTjJ2A+w///lP40GdThcQEMCyXUtLS05OTkJCws6dO0VXNNqzX375ZcSIEQ4O\nDpGRkcuXL2f75AYEBDQ0NLAJomxnvGuinUCw6zNmlgtfuXIlEQ0fPvzgwYPXr18vKyv785//\nzHHc559/LnfV1uJ23XvhhRfQOvPwwpMMrZOsqqpKqVQuXbo0JSXFwmyHFTpFNm3a1O3VnBUV\nFY6OjqL17UDw9NNPq1Sqr7/+mj3U6/UTJ04kol27dglzhGxnnxskItj1GfPLhW/fvt3JyUn4\nXH3QoEHG2/+Bme5t3boVrTMDLzzJ0LreeOGFF1QqVU1NjSXZDit0mjpw4AARpaammh6aM2cO\nO8+IbCdy69YtjuNeeuklYSQpKYmINm7cyPP8iRMnhPHY2Fh7OwMrQLDrG5YsF15fX79169Yl\nS5a8/fbbZv4HtEM9dg+tux288CRD63rp8uXLjo6Ozz//PM/zlmQ7ELl+/bpSqfT19f3tt99E\nhxYuXLhgwQI7vDisRxcuXCAiYWlr41S3bds2lUp15coVWQu0Cgh2fQPLhfcGuicZWicZWmeJ\n9vb2w4cP3+58VkxMjFKpZGEO2a5Hra2t2dnZO3furK+vZyNvvvkmEc2ZM8d4r+Hm5matVvu3\nv/3NDi8O61Fzc7NarY6Ojub/f6rjef7TTz8louPHj8taoFVAsOsbWC68N9A9ydA6ydC6HrH1\n6oho1KhRe/bsMY13V65cUavVzz77LHuIbGdGRUWFv78/O60/ePDgo0eP8jxvMBgef/xxIvL3\n98/Ozr527VpJSQlb2eTUqVNyl2yl5syZo1AoFixYYJzq+P8Fu7y8PBlrsxIIdn0Gy4X3Bron\nGVonGVpnXl5eHtvaValUEpGfn9++fftEH2SuWLHCwcGhoqKCPUS261Ztbe3QoUPd3d3XrFmT\nnJzs6urq6OiYm5vL83xnZ2dCQoKwhS4RqdXqzZs3y12ytWhpadm8efPixYs3bdrEfrW4ePGi\nRqMRXUfR2dk5Y8YMV1dXrEbEI9j10tWrV//97383NzfzPI/lwu+Icet4dO9OoHWSoXV36sMP\nP+Q4bsiQIa+++qqfnx8RPfDAAx999JHQnKampnvuuYedGmNYttuwYYNMJVujpKQkpVIp/ApR\nXl7u6ekpZDue50tLS5OSkubPn79y5coff/xRvkqtS1VV1dixY4XIe+jQITaemZnp6Ojo7Oz8\n3nvvnT17Njc3d9KkSUSUkZEhb8FWAsFOosbGxvDwcPZqc3JyioyMvHLlCpYLt0S3reN5Ht3r\nEVonGVpnufb29qioKCFzsGzn4eHx3Xff7dmzZ9SoUUQ0ZsyYgwcPskX/X3nlFYVCUVZWJnyH\nwsJCeUq3PuXl5bGxscuWLZs5c6bx+I8//ijKdiBy8+ZNf39/V1fXjIyMuro60Y/k+fPnR44c\nKWQ+V1fX/fv3y1WqtUGwk6K0tHTYsGEKhSIyMjIuLi4wMJCIPDw8vvrqKywXbp6Z1vE8j+6Z\ngdZJhtZZTriubsWKFcKgkO1KS0s7Ojp27drl6+tLRA899NChQ4d+/vlnFxeX+fPny1i21Xrr\nrbeIyM3Nbdq0aaJDyHbmbdy4UXQzRFtb24EDBxITE/Pz83me7+joyM/Pf+eddzIyMpqamuSr\n1Oog2EkRFhbGcdyXX34pjBw8eNDZ2dnFxaWyshLLhZthvnU8z6N7t4PWSYbWWUhIdTExMaJb\nMo2zHc/zer1+586d3t7e7ML/wMBAjuO+//57mQq3avHx8URkfCWiANnOjCeeeGLw4MHCw0OH\nDmm1WvbbF8dxOOtqBoLdnSkvLz9z5sz9998/Z84c0aHjx49zHPeHP/xBlsKsH1onGVonGVpn\nOTOpjhFlO/ZX0tPTR4wYwd5uP/jgg/4tecBg2S4gIOD3338XHWLZbu3atbIUZs2WLFlCRG+9\n9VZmZub06dOJyMfH5/3338/JyRk2bJiPj4/cBVovBLs7M336dBcXF1dX14ULF5oejY6OJiL8\n2tottE4ytE4ytM5CQqqbNGmSmeXTTLMdz/NtbW1btmx5++23+6XSgeHSpUvsw2ABy3ZBQUGm\n2e7atWv9WNqAUVNT4+rqKlwX+9prr7W0tLBDixcvViqV8pZnzRDs7kx9ff3o0aOJyNPT0/Sm\nuYyMDCLCnerdQuskQ+skQ+ssIaQ6lUpFPW1j1W22A8HPP/88e/ZsFkf+8pe/GMc4M9kOBMaZ\nuLq6+vXXX1+/fr2wpDPP87du3fLx8QkNDZWpwAEAwe6OCW8VCQkJokO7d+8mok8//VSWwqwf\nWicZWicZWmee8RnYb7/91t3dHdlOssbGRq1Wq1arFy9ezO7CFsU4ZDszzGRinudv3LjB/gwP\nD1coFMIWz2AKwU4K9lbBcVxKSopw2kKv148fP37w4MGNjY3ylmfN0DrJ0DrJ0LrbMb2urqSk\n5I6ynU6n669irV1XV9ef/vQnZ2fns2fPspGwsDAz2a6trU2mSq2R+Uycmprq6uoaHh7u6enJ\ncRxWSTQPwU4i4WOAKVOmpKWlbdq0ady4cRzHffjhh3KXZu3QOsnQOsnQum5dvXp1zJgxorsl\nLM925ifYm6KiIiJKS0tjD7u6uqZOnerj49NttktKSpKnSqvUYyb+4IMP2GvSz8/vyJEjshY7\nACDYSSe8VXAc5+npGRwcjA+HLYTWSYbWSYbWdev69eumd0tYmO3A2JYtW4joxIkT7OH69eud\nnJwqKyvZXvVBQUGNjY1ZWVlm7k2xW5Zk4q6uLnw8bCEEu15hbxUKhcKeVzSVBq2TDK2TDK2z\nHLKdhaqrq9kXp0+fJqI1a9bwPF9YWKhUKtPT09mhKVOmCCtgY1sOU8jEfQvBrrfwViEZWicZ\nWicZWmc5ZLsedXZ2+vn5JSYmsoe7d+/u6Ohob28fPXr07NmzhWlz586dPXv2xIkTcQt2t5CJ\n+xaCXR/AW4VkaJ1kaJ1kaJ3lkO3M++yzzziOM972iuf5nJwcIsrLy2MPm5ub3dzczpw5I0eB\nAwYycR9CsOsbwlvFxx9/LHctAwxaJxlaJxlaZzlku26Vl5evXLly//79gYGBokPbtm0jIrYn\nvcFgWLBgwb333issrgtMXV1dQUGBaHFmZOI+oSDoC1qtNj8/f8aMGZMnT5a7lgEGrZMMrZMM\nrbNcYGDgqVOn3N3deZ6XuxYrcujQoXXr1m3evPmee+4RHZo1a5ZarV66dOmSJUtCQkIOHDiw\ndetWZ2dnWeqUnV6vP3nypPFIY2PjvHnzvL29Z86c6enpGRYWVltbyw41NTURkU6nI6Kurq6Y\nmBilUjlhwoT+L3tgkztZAgCAtbPn1f5uJzY2logUCkVFRYXo0LFjx4YNG0ZEQ4cOZR/d2a1l\ny5YZX/bwyy+/eHt7cxwXERERFxcXGBhIRO7u7idPnuR5vqqqSq1WDxo06MUXX2RX133yySey\nlj8gIdgBAABIwbKdv79/c3Oz6FBnZ2dNTY3BYJClMOshuqR19erVHMd9+eWXwoSDBw86Ozu7\nubldunSJRybuCxyPT9cBAAAkiYuLe//992fMmPH555+7uLjIXY41unLlSmhoaE1Nza5du4qK\nipqamo4ePWo84fjx4/PmzZs2bVphYSERGQyG+vp6b29vhQJXi0mBYAcAACAdsl2PhGyn1Wof\neeSRvXv3iiZER0dnZGRUVVWxVcShNxCHAQAApEtLS4uNjS0sLJw7d25LS4vc5VgjdrvSyJEj\n6+rqTpw4YTAYRBMef/xxIiooKJCjOluDYAcAANArxtlOr9fLXY41Ytlu9OjRTU1N//rXv0RH\nW1tbiYhdXQe9pJS7AAAAgAEvLS2NiNzc3BwdHeWuxUqxbBcaGpqamuri4rJq1SqO44ioo6Nj\n06ZNbm5uxvtMgGS4xg4AAAD6Cbverrq6Ojg4ODo62sHBYfv27aWlpXv37l20aJHc1dkCBDsA\nAADoP0K2Yw9Hjx69bdu2WbNmyVuVzUCwAwAAgH7Fst3ly5dzcnLmzZsndzk2BTdPAAAAQL9i\n19v5+vpeuHBB7lpsDT6xAwAAABncuHFj8ODBcldhaxDsAAAAAGwETsUCAAAA2AgEOwAAAAAb\ngWAHAAAAYCMQ7AAAAABsBIIdAAAAgI1AsAMAAACwEQh2AAAAADYCwQ4A7FFycjLHcRzHFRYW\ndjshISGBTTh37lxv/qEZM2aMGDHCwskLFy7kOK69vb03/yIA2DMEOwCwa5mZmd2OZ2Vl9XMl\nAAC9h2AHAPZLo9FkZWWZbsBTVlZ28eJFbHYEAAMOgh0A2K958+Y1NDQUFRWJxrOysjiOmzt3\nrixVAQBIhmAHAPYrPDzcwcHh8OHDovGsrKzg4GCtVisav3HjxooVK8aPH6/RaCZNmhQfH9/a\n2mo84eLFi1FRUSNGjNBqtU899dTly5dF38FgMKSkpISEhGg0mpEjRy5fvrypqamvnxYA2C8E\nOwCwX8OGDQsNDRWdjf3pp59KS0ujoqJEk3U6XWBg4ObNmzUazdNPP83z/LvvvjtlypSWlhY2\noaioaPLkydnZ2X5+fo8++ug333wTEhJSX18vfAe9Xv/oo48mJiZ2dnZGR0f7+vpu2bIlODi4\nrq6uH54sANgDBDsAsGtRUVH19fXnz58XRtjtFBEREaKZa9asqa6uTktLO3PmzM6dO0tKSuLj\n48vKyjZu3MgmxMXFtbS0ZGdnnz59ev/+/RcuXPDz86utrRW+Q3p6ekFBwdq1a4uLi3fs2HH6\n9Ol9+/bV1dX94x//uPtPFADsAoIdANi1J598UqFQGJ+NzcrKmjhxoq+vr/G0jo6OXbt2jRs3\nbsWKFcJgcnKyp6fn9u3biaikpOT8+fORkZHh4eHsqEaj2bBhg/E32bBhg5+fX0JCgjDy7LPP\nhoSEHD169NatW3fhyQGA3VHKXQAAgJzuu+++mTNnZmZmrlu3johqa2tLSkpSU1NF02pra/V6\nfWhoKMdxwqBarZ46dWp2dvbNmzcrKyuJKCwszPhvTZ482cPDg3198+bNurq6kJCQjIwM4zlO\nTk56vb66ujogIOBuPEEAsCsIdgBg76Kiol5++eXi4uLJkyez5esiIyNFcxoaGojI09NTNO7l\n5cWOsnsg7r//ftEErVZ77do1ImLnZIuKikxvwiUi4UI9AIDewKlYALB3ERERCoWCXVqXlZUV\nEBDwwAMPiOawxKbT6UTjbMTLy4ttL2F6i+t///tf9sV9991HRMuWLeO7ExIS0vdPDADsD4Id\nANg7Ly+vqVOnHj58mK1pZ3o/LBH5+vqqVKqvv/7aeFCv1xcVFXl6emo0mgcffJCI8vLyjCfU\n1NQId7wOHTp06NChphuUrVu3LikpqS+fDwDYMQQ7AACaP39+TU3NG2+8wfO86XlYIlKpVM89\n99wPP/ywdetWYXDNmjUNDQ0xMTFENGHChODg4MOHD+fm5rKjbW1ty5cv7+rqEuYvXbr022+/\nTUlJEUb279//6quvVlVV3a0nBgB2BtfYAQBQREREXFzcnj17xowZ4+/v3+2cpKSkvLy8l19+\nOTMzc+zYsSUlJefOnXv44YdfeeUVNmHDhg1hYWFPPvnkY4895u3tnZ+fr9Pppk+fLixTHB8f\nf+TIkcTExOzs7KCgoIaGhmPHjg0fPvzdd9/tn6cJADYPn9gBAJBWqw0ODqbubpsQeHl5ff/9\n98uWLfv111/37dvX0dGxatWqc+fOOTs7swkhISHFxcURERGVlZXHjx8fP358QUHBuHHjhO+g\n0WiKi4vj4+N5nt+7d++FCxdefPHF4uLi4cOH3+0nCAB2gjPd/RoAAAAABiJ8YgcAAABgIxDs\nAAAAAGwEgh0AAACAjUCwAwAAALARCHYAAAAANgLBDgAAAMBGINgBAAAA2AgEOwAAAAAbgWAH\nAAAAYCMQ7AAAAABsBIIdAAAAgI1AsAMAAACwEQh2AAAAADYCwQ4AAADARvwfyRlen5KtJR8A\nAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "metrics <- c(\"MCC\", \"Precision\")\n",
    "models <- c(subsets, \"baseline\")\n",
    "# Summarize the performance of all the models in a data frame\n",
    "dat <- data.frame()\n",
    "for (i in 1:length(models)){\n",
    "    perf <- read.table(paste0(\"Response_to_odorants/Cross_validation/\", models[i], \".tsv\"),\n",
    "                       sep = \"\\t\", row.names = 1)\n",
    "    perf <- subset(perf[1:5, ], select = metrics)\n",
    "    dat <- rbind(dat, merge(models[i], perf))\n",
    "}\n",
    "names(dat) <- c(\"Model\", metrics)\n",
    "\n",
    "col_palette <- c('orange', 'cyan3', 'purple4', 'red', 'gray30', 'gray60', 'black')\n",
    "\n",
    "# Plot the MCC\n",
    "ggerrorplot(dat, x = \"Model\", y = \"MCC\", color = \"Model\", palette = col_palette,\n",
    "            add = \"mean\", add.params = list(size = 0.5),\n",
    "            error.plot = \"errorbar\", width = 0.2, legend = \"none\") + \n",
    "coord_cartesian(ylim = c(-0.05, 0.8)) + rotate_x_text(45) +\n",
    "stat_compare_means(label = \"p.signif\", label.y = 0.75, method = \"t.test\", ref.group = \"random\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation:  \n",
    "All the 5 models performed significantly better than the control (*random*).  \n",
    "However, the differences among the 5 models are not statistically significant.  \n",
    "We choose the model ***poc60***, which is a good compromise between performance and speed.\n",
    "\n",
    "*Note:* The data contain 26% responsive (positive) OR-odorant pairs. Therefore, the ~26% `hit rate` of the control model is equivalent to a random prediction.  \n",
    "\n",
    "## 3. Prediction of new ORs  \n",
    "- Rebuild the best model above (***poc60***), with all the training data (without splitting into training & validation sets).  \n",
    "- Use the model to predict the response probabilities of 416 query ORs paired with the 4 odorants (416 $\\times$ 4 = 1664 pairs).  \n",
    "- Repeat the prediction 3 times and rank the pairs by the average response probability.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the training data columns belonging to poc60, as well as the odorant columns and the response label \n",
    "subset <- \"poc60\"\n",
    "ResCols <- paste0(\"Res\", as.vector(outer(get(subset), c(\"Comp\", \"Polar\", \"Vol\"), paste0)))\n",
    "datTrain <- data[, names(data) %in% c(ResCols, names(datOdor), \"Label\")]\n",
    "\n",
    "# Generate the test set of OR-odorant pairs in the same format\n",
    "queryORs <- readLines(\"Response_to_odorants/Data/Query_ORs.txt\")\n",
    "queryORs <- datOR[rownames(datOR) %in% queryORs, ]\n",
    "\n",
    "testData <- data.frame()\n",
    "for (cid in testCIDs){\n",
    "  test <- merge(queryORs, datOdor[datOdor$CID == cid, -1])\n",
    "  rownames(test) <- paste(rownames(queryORs), cid, sep = \"_\")\n",
    "  testData <- rbind(testData, test)\n",
    "}\n",
    "testSet <- testData[, colnames(testData) %in% colnames(datTrain)]\n",
    "\n",
    "# Center & scale data\n",
    "preProc <- preProcess(datTrain, method = c(\"center\", \"scale\"))\n",
    "trainingProc <- predict(preProc, datTrain)\n",
    "testProc <- predict(preProc, testSet)\n",
    "\n",
    "# An empty data frame to store the prediction\n",
    "prediction <- as.data.frame(matrix(nrow = nrow(testProc)), ncol=3)\n",
    "rownames(prediction) <- rownames(testProc)\n",
    "\n",
    "# Parameter tuning using nested 5-fold cross validation\n",
    "control <- trainControl(classProbs = TRUE, \n",
    "                        method = \"cv\", \n",
    "                        number = 5,\n",
    "                        summaryFunction = prSummary)\n",
    "\n",
    "# Repeat 3 times the training - prediction process\n",
    "for (i in 1:3){\n",
    "    # Model training using the same parameter tuning procedure as above\n",
    "    grid <- expand.grid(.mtry = seq(ncol(trainingProc)/2, ncol(trainingProc)-1, length.out= 5))\n",
    "    mod_final <- train(data=trainingProc, Label ~ . ,\n",
    "                       method=\"rf\", ntree = 2001,\n",
    "                       metric=\"Precision\",\n",
    "                       tuneGrid = grid,\n",
    "                       trControl=control)\n",
    "    \n",
    "    # Predict the response probabilities for the test set\n",
    "    probability <- predict(mod_final, newdata = testProc, type = \"prob\")\n",
    "    prediction[, i] <- probability$positive\n",
    "}\n",
    "\n",
    "colnames(prediction) <- paste0('Prediction', 1:3)\n",
    "Mean <- rowMeans(prediction)\n",
    "SD <- apply(prediction, 1, sd)\n",
    "prediction <- cbind(prediction, Mean, SD)\n",
    "prediction <- round(prediction, 2)\n",
    "\n",
    "# Rank by the predicted response probability\n",
    "prediction <- prediction[order(prediction$Mean, decreasing = TRUE), ]\n",
    "\n",
    "# Output to file\n",
    "write.table(prediction, paste0(\"Response_to_odorants/Prediction/\", subset, \".tsv\"), sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List the predicted responsive OR-odorant pairs and the ORs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicted responsive pairs: \n",
      "               Replicate1 Replicate2 Replicate3 Mean   SD\n",
      "Olfr885_7410         0.93       0.90       0.91 0.91 0.02\n",
      "Olfr885_323          0.89       0.87       0.88 0.88 0.01\n",
      "Olfr205_7410         0.87       0.84       0.84 0.85 0.02\n",
      "hOR8B2_439570        0.86       0.85       0.82 0.84 0.02\n",
      "Olfr194_7410         0.83       0.84       0.84 0.84 0.00\n",
      "Olfr913_7410         0.84       0.81       0.82 0.82 0.01\n",
      "Olfr1057_7410        0.81       0.81       0.80 0.81 0.01\n",
      "Olfr251_7410         0.81       0.80       0.79 0.80 0.01\n",
      "Olfr160_7410         0.80       0.80       0.80 0.80 0.00\n",
      "Olfr1057_323         0.81       0.79       0.79 0.80 0.01\n",
      "Olfr44_7410          0.78       0.79       0.80 0.79 0.01\n",
      "Olfr199_7410         0.79       0.79       0.78 0.79 0.01\n",
      "Olfr913_323          0.80       0.77       0.78 0.78 0.02\n",
      "hOR1E3_7410          0.78       0.76       0.78 0.77 0.01\n",
      "Olfr251_323          0.78       0.77       0.77 0.77 0.00\n",
      "Olfr1057_68110       0.77       0.79       0.75 0.77 0.02\n",
      "Olfr747_7410         0.75       0.75       0.74 0.75 0.01\n",
      "Olfr205_323          0.78       0.75       0.73 0.75 0.02\n",
      "Olfr747_323          0.76       0.76       0.75 0.75 0.01\n",
      "Olfr194_323          0.74       0.75       0.73 0.74 0.01\n",
      "Olfr885_68110        0.74       0.74       0.73 0.74 0.01\n",
      "Olfr147_7410         0.74       0.72       0.71 0.72 0.02\n",
      "Olfr199_323          0.70       0.72       0.72 0.72 0.01\n",
      "hOR8B2_7410          0.70       0.70       0.71 0.70 0.01\n",
      "hOR1E3_323           0.70       0.69       0.72 0.70 0.02\n",
      "Olfr44_323           0.68       0.69       0.70 0.69 0.01\n",
      "hOR8B2_323           0.68       0.67       0.68 0.68 0.00\n",
      "Olfr147_323          0.69       0.67       0.67 0.68 0.01\n",
      "Olfr908_7410         0.69       0.67       0.67 0.67 0.01\n",
      "Olfr160_323          0.67       0.66       0.68 0.67 0.01\n",
      "Olfr908_323          0.67       0.65       0.66 0.66 0.01\n",
      "Olfr251_68110        0.64       0.63       0.62 0.63 0.01\n",
      "Olfr893_7410         0.66       0.64       0.60 0.63 0.03\n",
      "Olfr893_323          0.66       0.62       0.60 0.63 0.03\n",
      "Olfr1487_7410        0.59       0.65       0.63 0.62 0.03\n",
      "Olfr147_439570       0.63       0.61       0.60 0.61 0.02\n",
      "hOR8B2_68110         0.61       0.61       0.61 0.61 0.00\n",
      "Olfr944_7410         0.59       0.60       0.60 0.60 0.01\n",
      "Olfr141_7410         0.56       0.59       0.59 0.58 0.02\n",
      "Olfr147_68110        0.60       0.56       0.57 0.58 0.02\n",
      "hOR1E3_439570        0.55       0.53       0.59 0.56 0.03\n",
      "Olfr885_439570       0.55       0.55       0.57 0.56 0.01\n",
      "Olfr944_323          0.55       0.56       0.57 0.56 0.01\n",
      "Olfr44_68110         0.57       0.56       0.55 0.56 0.01\n",
      "Olfr160_68110        0.57       0.54       0.56 0.56 0.02\n",
      "Olfr194_439570       0.53       0.56       0.57 0.55 0.02\n",
      "Olfr205_439570       0.53       0.54       0.56 0.54 0.01\n",
      "Olfr1471_7410        0.52       0.58       0.53 0.54 0.03\n",
      "Olfr1487_323         0.51       0.56       0.56 0.54 0.03\n",
      "Olfr141_323          0.53       0.54       0.56 0.54 0.01\n",
      "Olfr913_68110        0.55       0.53       0.54 0.54 0.01\n",
      "Olfr194_68110        0.54       0.54       0.53 0.54 0.01\n",
      "Olfr987_439570       0.52       0.51       0.54 0.53 0.02\n",
      "Olfr988_439570       0.52       0.51       0.55 0.53 0.02\n",
      "Olfr951_7410         0.52       0.52       0.55 0.53 0.01\n",
      "Olfr205_68110        0.54       0.53       0.52 0.53 0.01\n",
      "Olfr338_7410         0.54       0.51       0.50 0.52 0.02\n",
      "hOR1E3_68110         0.54       0.51       0.51 0.52 0.02\n",
      "Olfr199_439570       0.49       0.51       0.54 0.51 0.02\n",
      "\n",
      "Predicted responsive ORs: \n",
      " hOR1E3 hOR8B2 Olfr1057 Olfr141 Olfr147 Olfr1471 Olfr1487 Olfr160 Olfr194 Olfr199 Olfr205 Olfr251 Olfr338 Olfr44 Olfr747 Olfr885 Olfr893 Olfr908 Olfr913 Olfr944 Olfr951 Olfr987 Olfr988"
     ]
    }
   ],
   "source": [
    "# Pairs with response probability > 0.5\n",
    "subset <- \"poc60\"\n",
    "prediction <- read.table(paste0(\"Response_to_odorants/Prediction/\", subset, \".tsv\"), stringsAsFactors = F)\n",
    "topPairs <- prediction[prediction$Mean > 0.5, ]\n",
    "cat(\"Predicted responsive pairs: \\n\")\n",
    "print(topPairs)\n",
    "\n",
    "topORs <- unique(gsub(\"_[0-9].*\", \"\", rownames(topPairs)))\n",
    "cat(\"\\nPredicted responsive ORs: \\n\", sort(topORs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Evaluate the prediction with *in vitro* data  \n",
    "To evaluate the model performance on both classes (positive and negative), a test set of ORs underwent *in vitro* functional assays with the 4 odorants.  \n",
    "The test set contained:\n",
    "- all the predicted responsive ORs (the positive class),  \n",
    "- randomly picked ORs from the negative class.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>Replicate1</th><th scope=col>Replicate2</th><th scope=col>Replicate3</th><th scope=col>Mean</th><th scope=col>SD</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Olfr885_7410</th><td>0.93</td><td>0.90</td><td>0.91</td><td>0.91</td><td>0.02</td></tr>\n",
       "\t<tr><th scope=row>Olfr885_323</th><td>0.89</td><td>0.87</td><td>0.88</td><td>0.88</td><td>0.01</td></tr>\n",
       "\t<tr><th scope=row>Olfr205_7410</th><td>0.87</td><td>0.84</td><td>0.84</td><td>0.85</td><td>0.02</td></tr>\n",
       "\t<tr><th scope=row>hOR8B2_439570</th><td>0.86</td><td>0.85</td><td>0.82</td><td>0.84</td><td>0.02</td></tr>\n",
       "\t<tr><th scope=row>Olfr194_7410</th><td>0.83</td><td>0.84</td><td>0.84</td><td>0.84</td><td>0.00</td></tr>\n",
       "\t<tr><th scope=row>Olfr913_7410</th><td>0.84</td><td>0.81</td><td>0.82</td><td>0.82</td><td>0.01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & Replicate1 & Replicate2 & Replicate3 & Mean & SD\\\\\n",
       "\\hline\n",
       "\tOlfr885\\_7410 & 0.93 & 0.90 & 0.91 & 0.91 & 0.02\\\\\n",
       "\tOlfr885\\_323 & 0.89 & 0.87 & 0.88 & 0.88 & 0.01\\\\\n",
       "\tOlfr205\\_7410 & 0.87 & 0.84 & 0.84 & 0.85 & 0.02\\\\\n",
       "\thOR8B2\\_439570 & 0.86 & 0.85 & 0.82 & 0.84 & 0.02\\\\\n",
       "\tOlfr194\\_7410 & 0.83 & 0.84 & 0.84 & 0.84 & 0.00\\\\\n",
       "\tOlfr913\\_7410 & 0.84 & 0.81 & 0.82 & 0.82 & 0.01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | Replicate1 | Replicate2 | Replicate3 | Mean | SD |\n",
       "|---|---|---|---|---|---|\n",
       "| Olfr885_7410 | 0.93 | 0.90 | 0.91 | 0.91 | 0.02 |\n",
       "| Olfr885_323 | 0.89 | 0.87 | 0.88 | 0.88 | 0.01 |\n",
       "| Olfr205_7410 | 0.87 | 0.84 | 0.84 | 0.85 | 0.02 |\n",
       "| hOR8B2_439570 | 0.86 | 0.85 | 0.82 | 0.84 | 0.02 |\n",
       "| Olfr194_7410 | 0.83 | 0.84 | 0.84 | 0.84 | 0.00 |\n",
       "| Olfr913_7410 | 0.84 | 0.81 | 0.82 | 0.82 | 0.01 |\n",
       "\n"
      ],
      "text/plain": [
       "              Replicate1 Replicate2 Replicate3 Mean SD  \n",
       "Olfr885_7410  0.93       0.90       0.91       0.91 0.02\n",
       "Olfr885_323   0.89       0.87       0.88       0.88 0.01\n",
       "Olfr205_7410  0.87       0.84       0.84       0.85 0.02\n",
       "hOR8B2_439570 0.86       0.85       0.82       0.84 0.02\n",
       "Olfr194_7410  0.83       0.84       0.84       0.84 0.00\n",
       "Olfr913_7410  0.84       0.81       0.82       0.82 0.01"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "poc60 Model performance on the test set:"
     ]
    },
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>MCC</dt>\n",
       "\t\t<dd>0.37</dd>\n",
       "\t<dt>Precision</dt>\n",
       "\t\t<dd>0.44</dd>\n",
       "\t<dt>Recall</dt>\n",
       "\t\t<dd>0.53</dd>\n",
       "\t<dt>Specificity</dt>\n",
       "\t\t<dd>0.86</dd>\n",
       "\t<dt>TP</dt>\n",
       "\t\t<dd>26</dd>\n",
       "\t<dt>TN</dt>\n",
       "\t\t<dd>210</dd>\n",
       "\t<dt>FP</dt>\n",
       "\t\t<dd>33</dd>\n",
       "\t<dt>FN</dt>\n",
       "\t\t<dd>23</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[MCC] 0.37\n",
       "\\item[Precision] 0.44\n",
       "\\item[Recall] 0.53\n",
       "\\item[Specificity] 0.86\n",
       "\\item[TP] 26\n",
       "\\item[TN] 210\n",
       "\\item[FP] 33\n",
       "\\item[FN] 23\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "MCC\n",
       ":   0.37Precision\n",
       ":   0.44Recall\n",
       ":   0.53Specificity\n",
       ":   0.86TP\n",
       ":   26TN\n",
       ":   210FP\n",
       ":   33FN\n",
       ":   23\n",
       "\n"
      ],
      "text/plain": [
       "        MCC   Precision      Recall Specificity          TP          TN \n",
       "       0.37        0.44        0.53        0.86       26.00      210.00 \n",
       "         FP          FN \n",
       "      33.00       23.00 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Model performance on R-(-)-Carvone \n",
      "        MCC   Precision      Recall Specificity          TP          TN \n",
      "       0.40        0.56        0.42        0.93        5.00       57.00 \n",
      "         FP          FN \n",
      "       4.00        7.00 \n",
      "\n",
      "Model performance on acetophenone \n",
      "        MCC   Precision      Recall Specificity          TP          TN \n",
      "       0.32        0.33        0.64        0.77        7.00       48.00 \n",
      "         FP          FN \n",
      "      14.00        4.00 \n",
      "\n",
      "Model performance on coumarin \n",
      "        MCC   Precision      Recall Specificity          TP          TN \n",
      "       0.34        0.44        0.53        0.83        8.00       48.00 \n",
      "         FP          FN \n",
      "      10.00        7.00 \n",
      "\n",
      "Model performance on 4-chromanone \n",
      "        MCC   Precision      Recall Specificity          TP          TN \n",
      "       0.46        0.55        0.55        0.92        6.00       57.00 \n",
      "         FP          FN \n",
      "       5.00        5.00 \n"
     ]
    }
   ],
   "source": [
    "subset <- \"poc60\"\n",
    "# Read the in vitro response data\n",
    "newPairs <- read.csv(paste0(\"Response_to_odorants/Data/TestSet_OR-odorant_pairs/TestSet_\", subset, \".csv\"),\n",
    "                     stringsAsFactors = FALSE)\n",
    "\n",
    "# Merge OR-odorant pairs\n",
    "pos_pairs <- as.character()\n",
    "neg_pairs <- as.character()\n",
    "for (cid in testCIDs){\n",
    "    # responsive ORs & pairs\n",
    "    pos <- newPairs$Gene[newPairs$CID == cid & newPairs$Responsive == 1]\n",
    "    pos_pairs <- c(pos_pairs, paste0(pos, \"_\", cid))\n",
    "    # non-responsive ORs & pairs\n",
    "    neg <- newPairs$Gene[newPairs$CID == cid & newPairs$Responsive == 0]\n",
    "    neg_pairs <- c(neg_pairs, paste0(neg, \"_\", cid))\n",
    "}\n",
    "in_vitro <- rbind(cbind(pos_pairs, \"positive\"), cbind(neg_pairs, \"negative\"))\n",
    "colnames(in_vitro) = c('Pairs', 'Label')\n",
    "\n",
    "\n",
    "prediction <- read.csv(paste0(\"Response_to_odorants/Prediction/\", subset, \".tsv\"), sep = \"\\t\")\n",
    "head(prediction)\n",
    "# Extract prediction on the pairs tested in vitro\n",
    "testSet <- prediction[in_vitro[, \"Pairs\"], 4:5]\n",
    "testSet$Predicted <- factor(ifelse(testSet$Mean > 0.5 , \"positive\", \"negative\"),\n",
    "                           levels = c(\"positive\", \"negative\"))\n",
    "testSet$in_vitro <- factor(in_vitro[, \"Label\"], levels = c(\"positive\", \"negative\"))\n",
    "\n",
    "# Evaluate the performance on the whole test set\n",
    "perfSum <- getMetrics(testSet$Predicted, testSet$in_vitro, positive = \"positive\")\n",
    "cat(subset, \"Model performance on the test set:\")\n",
    "perfSum\n",
    "\n",
    "# Evaluate the performance for each odorant\n",
    "for (cid in testCIDs){\n",
    "    test <- testSet[grep(paste0(\"_\", cid), rownames(testSet)), ]\n",
    "    perf <- getMetrics(test$Predicted, test$in_vitro, positive = \"positive\")\n",
    "    cat(\"\\nModel performance on\", allOdor$Odorant[match(cid, allOdor$CID)], \"\\n\")\n",
    "    print(perf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
